{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Rosalind Problems - Bioinformatics Stronghold"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Counting DNA Nucleotides (nt)\n",
    "\n",
    "The nucleus of any living cell consists of macromolecules called chromatin. One class of marcomolecules in the chormatin are nucleic acids. Nucleic acids are polygmers - repeating chains of small, similarly strucutred molecules called monomers (stands.)\n",
    "\n",
    "The **nucleotide (nt)** is the unit of stand length and is composed of three parts: a sugar molecule, a negatively charged phosphate ion, and a compound called a nucleobase (base.) THe sugar of one nucleotide bonds to the the phosphate of the next nucleotide in the chain, forming a sugar-phosphate backbone.\n",
    "\n",
    "Nucleotides of a specific type of nucleic acid always contains the same sugar and phosphate molecule; thus they only differ by their choice of base. Therefore a strand of nucleic acid can be differentiated from another based only on the order of its bases - this ordering defines a nucleic acid's **primary structure**. \n",
    "\n",
    "For a strand of **deoxyribose nucleic acid (DNA)**, the four necleobases are molecules called **adenine (A)**, **cytosine (C)**, **guanine (G)**, and **thymine (T)**."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem: \n",
    "\n",
    "Given a string over the alphabet {A,C,G,T}, count the number of times each symbol occurs in the string."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_values([247, 244, 242, 242])\n"
     ]
    }
   ],
   "source": [
    "input_string = \"ATTCGTAGGCATCGCTCCTGCCCTGTGGTTACGTTCGCACCGGGTGTCAGGTCTGAAATTATGAGAATTTTACACTCCTAATGTTGGGCGACTGAACTGATTTTGAAGTCAGGCGGTAGTATACTGAGTCAGGGTCGGAGAGTGAGAGATGTATGAGCGTGGTTTGCTCCGCTAGCCCAGGCCGGCGAGTGCTGATGCGTAGCTGTATCTACAATTCGTCAGCTTTTGGCGCAGTAATGAATATCAGAGTGCCTAAGCAGCTCTTGCCCTACGGCTAATCGAATCTTCAGATTCAAAGGGGTACACTAGCTGTTGCCAAACCGTCGGAGACGCCACCTCACGAACTCACATCATATCAGTCGAAGATGGTACGTTAGAGCCCATGTGCCTATGGAGGCATGGTATATAGCCTGGCTGCGGTGTAATGAAGTCACCTGGCCAACCTGGGGCACTTGTCTCGTGACATAGTCTTATAGGGGTGCCGCAAGAAGGTCCCACCAAAAATCACGGATACGCGTTCGCCCGAAGGCCCTTGGAAACCTTCCAGCTCAAACTTGAGAGATACGTAGTTTTCCCACCCGTCACAAGAGGTGAGATTACAACAAAGTATGTCTGCTCGTCCCTGTTAGCTATAGCAGCTCGTATAACAACATTTCCCGACTGCGCGTCGCTGATCGAACAATAGGCTAGAGTTACAGAGCCTATCACAGGCCGAAGAGGACAAGCTTCGCTGGTACCTATAAAGGGAAAACCCCAGATTAATCCCGCCTTATCGTCCTAAGTAGTCGCGGTCAATATGGTACTATCCGACACCCTTTGATTGTGAGTCTTCACTTACCTATTCCGGATTTCCTCTCCCAGTAGGTTATAGTTCATGACGTCTTTTAGCACGGGATCACACGTGGCATGGATTAAATAGGATACACCTGGTGCCATGCAAAGTCCACGCTTAAGCCGAAGAACCCCGTAAAGCAC\"\n",
    "\n",
    "# Create dictonary for each character, scan through string, counting occurences.\n",
    "dict = {'A':0, 'C':0, 'G':0, 'T':0}\n",
    "\n",
    "for char in input_string:\n",
    "    count = dict[char]\n",
    "    dict[char] = count + 1\n",
    "    \n",
    "print(dict.values())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## The Second Nucleic Acid (RNA)\n",
    "\n",
    "Along side DNA in the chromatin a molecule with a different sugar can ribose known as **ribose nucleic acid (RNA)**. RNA differs from DNA as in it contains a base called **uracil** instread of thymine. \n",
    "\n",
    "The primary structure of DNA and RNA is similar because DNA serves as the blueprint for the creation of special kinds of RNA molecule called **messenger RNA (mRNA)**. mRNA is created during RNA transcription, during which a strand of DNA is used as a template for constrcuting a strand of RNA by compy nucleotides one at a time, where uracil is used in place of thymine."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem: \n",
    "\n",
    "Given a DNA string t, convert t to RNA by replacing all thymine (T) with uracil (U)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AAAACGUUGCCCUCCUCCCAGGUUCUGGUAAACGUGCACAGCGUCCCCUGCCUUUCAUAUCGUACUUCUAGGCCCUCGUAAGGAGUCCGGUAUGUAAUAUCGAUCACGAUCGCCUAUGAGGUGCGGGAAUUGCCAAAUCACUGGCCUAGUUGACUCAGUCACCCUCUUUUGCAAGGCCUUUAGCGAGGAAACCUGCACCGCCGAGGUACGCAAUUGCUACAUGCUACCCAAGAAGGGCAUUCAGCACGCGAGAGCGAUACUGUACUCGAGCACCACCAUCACGAUAGAAAGGAUGGUUGUAUUCACACGCGGUUAGGUGGAACCUUGAUAAAUGUGCGCGGAUGCGAAUGUGCCAAGUUGCCUGCUCUUUCUUAAGGUUCCGGUCUGACGGGAACACAAGGCGAGUAAGAGGGUCUUAGCAAGGGGUGCGCCUUAGCGACUACUAUUAGCUUGCACACAAACAAGCGAGUAGUAUGGCCUCGCUCGUGUUCAGCUCGGUUUGCGCGUGACCAUUGCAUUUUUUAAAUUGAAAAACCAGUGGUGCUGCUCACUAGCAUUAUCCGUUUGCUAGCCCUUCCAAUGUAACAAUGCAGUUGGGCUGAAGAGAGACCAGUACGCCCAGCUAUUCUAGAAUCCACCUCUGGAAGAACAACUGAGUAUUCUCGAAGUUUACCAGCAUCCCUCAGUAAAAUUUCAACUAUGAUGUUGUCAUCCAAUUGCUCCCGAGAAAAUUCGACGUGUAGCCCGAUGAAAGGUAAGUCGAGUGGGAUACCUCUAAUGAUAGUGCAGAGAAUGGCCCGUAGCCUCGGCACUAGAGCCCGUUGCCACUAAGAGCCAAGAAAAGCCCUCGUGAGCAGAGCCUCCGUGCAGUCUCUCUGGCCGGCAAUGCAAGGACAGGCUGGUUGC\n"
     ]
    }
   ],
   "source": [
    "t = \"AAAACGTTGCCCTCCTCCCAGGTTCTGGTAAACGTGCACAGCGTCCCCTGCCTTTCATATCGTACTTCTAGGCCCTCGTAAGGAGTCCGGTATGTAATATCGATCACGATCGCCTATGAGGTGCGGGAATTGCCAAATCACTGGCCTAGTTGACTCAGTCACCCTCTTTTGCAAGGCCTTTAGCGAGGAAACCTGCACCGCCGAGGTACGCAATTGCTACATGCTACCCAAGAAGGGCATTCAGCACGCGAGAGCGATACTGTACTCGAGCACCACCATCACGATAGAAAGGATGGTTGTATTCACACGCGGTTAGGTGGAACCTTGATAAATGTGCGCGGATGCGAATGTGCCAAGTTGCCTGCTCTTTCTTAAGGTTCCGGTCTGACGGGAACACAAGGCGAGTAAGAGGGTCTTAGCAAGGGGTGCGCCTTAGCGACTACTATTAGCTTGCACACAAACAAGCGAGTAGTATGGCCTCGCTCGTGTTCAGCTCGGTTTGCGCGTGACCATTGCATTTTTTAAATTGAAAAACCAGTGGTGCTGCTCACTAGCATTATCCGTTTGCTAGCCCTTCCAATGTAACAATGCAGTTGGGCTGAAGAGAGACCAGTACGCCCAGCTATTCTAGAATCCACCTCTGGAAGAACAACTGAGTATTCTCGAAGTTTACCAGCATCCCTCAGTAAAATTTCAACTATGATGTTGTCATCCAATTGCTCCCGAGAAAATTCGACGTGTAGCCCGATGAAAGGTAAGTCGAGTGGGATACCTCTAATGATAGTGCAGAGAATGGCCCGTAGCCTCGGCACTAGAGCCCGTTGCCACTAAGAGCCAAGAAAAGCCCTCGTGAGCAGAGCCTCCGTGCAGTCTCTCTGGCCGGCAATGCAAGGACAGGCTGGTTGC\"\n",
    "print(t.replace('T', 'U'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## The Secondary and Tertiary Structures of DNA\n",
    "\n",
    "The primary structure of a nucleic acid is determined by the ordering of its bases, yet it does not describe the large, 3D shape of the molecule. In 1953, the following structure for DNA was proposed:\n",
    "\n",
    "1) DNA is composed of two strings, running in opposite directions. \n",
    "\n",
    "2) Each base bonds to a base in the opposite strand: A-T, C-G (always)\n",
    "\n",
    "3) The two strands are twisted together into a long spiral double helix. \n",
    "\n",
    "\n",
    "1) and 2) compose the **secondary structure** of DNA. 3) Describes the **tertiary structure**.\n",
    "\n",
    "Note that the **complement** of a base is the base in which it bonds to. The bonding of two complementary bases is called a **base pair (bp)**. Thus the length of a DNA molecule is commonly given in bp rather than nt; we can determine the other strand by taking the complement of the first (running in opposite direction.)\n",
    "\n",
    "Example:\n",
    "\n",
    "'AAAACCCGGT' <-> 'ACCGGGTTTT'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TATAAACCGTGGGAAACGTCGATCCGCAACGACGAAAACTTTAGAAGGATCCTTATCTAATAGTGGTCCACGTTCCCCCTGTGCATTAACATAAAAAGAACCCTGACGTGATCCTCTAAACCTGGTACTACATTTTCACAGCTGGTTACACCGAGTGCTTAGCTTACGTCAGTACACGGATGAAAGGCACTGTCCTATACCTATCTTGAGTAGAGGGCAAGAATGGAAATTATACTCCGTTAGCTAAGGTATCTCTATAAGACATTAGGTTGTACCAAAAATGCAGCATTCAGGCGGGGTGCAGGGCTTGCCGGGGAACTGCTCGGATGGCTTGTAAGACAACACCCGCTCGAAGGTTCATGCAAACCTTACAATAACCGGAGACTCAGCTCGAAGAATTGCCCAATTTCGCCCCCTACCGCTGATTTGCGGCGATATGGGCCCTTCATATAATACAGAGTCTGCGCTACGAGACCTCAGTAGCCATCCCTAAATCGGTGGAGCGGCTTCTGACATTGGATCTCGAGCACTCTAGTCGACCGTTACTTTCAGAGCAATGCCTAAAAATGTCAAGCGCCGATGCGAAAGGATCCCATACACTGTAGTGAGTAGGTATGGACCTGGACGATACTAACGCTAGGCAGTCCCCTCAAGGGGCGCACCAGATCGCGTTGTCGTGAAGCGGCTAAGGGTTCCCCAGCCTTTACGAGACACAAAATGGCTGCCTAGAGTCATGTGGCCCGTTCTCTATCCCCAACTATCGTGTAAGATCGAGGAAAGTTGGGGTACGGCGGCGCCTCCTACGTGGGCGCCTCGTAACGATAGAGCCCCTGTCTGGAGTT\n"
     ]
    }
   ],
   "source": [
    "s = \"AACTCCAGACAGGGGCTCTATCGTTACGAGGCGCCCACGTAGGAGGCGCCGCCGTACCCCAACTTTCCTCGATCTTACACGATAGTTGGGGATAGAGAACGGGCCACATGACTCTAGGCAGCCATTTTGTGTCTCGTAAAGGCTGGGGAACCCTTAGCCGCTTCACGACAACGCGATCTGGTGCGCCCCTTGAGGGGACTGCCTAGCGTTAGTATCGTCCAGGTCCATACCTACTCACTACAGTGTATGGGATCCTTTCGCATCGGCGCTTGACATTTTTAGGCATTGCTCTGAAAGTAACGGTCGACTAGAGTGCTCGAGATCCAATGTCAGAAGCCGCTCCACCGATTTAGGGATGGCTACTGAGGTCTCGTAGCGCAGACTCTGTATTATATGAAGGGCCCATATCGCCGCAAATCAGCGGTAGGGGGCGAAATTGGGCAATTCTTCGAGCTGAGTCTCCGGTTATTGTAAGGTTTGCATGAACCTTCGAGCGGGTGTTGTCTTACAAGCCATCCGAGCAGTTCCCCGGCAAGCCCTGCACCCCGCCTGAATGCTGCATTTTTGGTACAACCTAATGTCTTATAGAGATACCTTAGCTAACGGAGTATAATTTCCATTCTTGCCCTCTACTCAAGATAGGTATAGGACAGTGCCTTTCATCCGTGTACTGACGTAAGCTAAGCACTCGGTGTAACCAGCTGTGAAAATGTAGTACCAGGTTTAGAGGATCACGTCAGGGTTCTTTTTATGTTAATGCACAGGGGGAACGTGGACCACTATTAGATAAGGATCCTTCTAAAGTTTTCGTCGTTGCGGATCGACGTTTCCCACGGTTTATA\"\n",
    "s_c = \"\" # complement of s. \n",
    "\n",
    "for char in s:\n",
    "    if char == 'A':\n",
    "        s_c += 'T'\n",
    "    elif char == 'T':\n",
    "        s_c += 'A'\n",
    "    elif char == 'C':\n",
    "        s_c += 'G'\n",
    "    elif char == 'G':\n",
    "        s_c += 'C'\n",
    "        \n",
    "print(s_c[::-1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Rabbits and Recurrences\n",
    "\n",
    "We begin with a pair of rabbits - one male and one female. Each month such a pair mates, and one month later k pairs are produced (one male, one female.)\n",
    "After a month of being born, rabbits reach reproductive age. Rabbits never stop reporducing or dying. \n",
    "\n",
    "Given n months and value k, what is the the total number of rabbit pairs after n months?\n",
    "\n",
    "e.g. n=5, k=3 -> 19 pairs. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Recurrence relationship:\n",
    "\n",
    "Month m=0, there is one pair of mating age, 0 new born. Total: 1\n",
    "Month m=1, there is one pair of mating age, k pairs of new borns. Total: k+1\n",
    "Month m=2, there are k+1 pairs of mating age, k pairs of new borns: Total: 2k+1\n",
    "\n",
    "In general, for the ith month suppose there are x,y pairs of mating age, new born respectively. Then for month i+1, we have\n",
    "x+y,k(x) respectvely. \n",
    "\n",
    "Let $F_i$ denote the number of pairs of rabbits on the $i^{th}$ month. Then for $i=0, i=1$, we have $F_0=1$, $F_1=k+1$, and $F_i = F_{i-1} + kF_{i-2}$ for $i \\geq 2$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def recurrence(m,k):\n",
    "    if m == 0:\n",
    "        return 0\n",
    "    elif m == 1:\n",
    "        return 1\n",
    "    else:\n",
    "        return recurrence(m-1, k) + k * recurrence(m-2, k) \n",
    "# See a bottom up solution."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3048504677680\n"
     ]
    }
   ],
   "source": [
    "m = 36\n",
    "k = 3\n",
    "print(recurrence(m, k))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GC-Content\n",
    "\n",
    "99.9% of the 3.2 billion base pairs in a human genome are common to almost all other humans (barring those with mahor genetic defects.) For any species, the average case genome is very similar.\n",
    "\n",
    "For DNA, this means that the frequencey of bases will be similar for memebers of the same species. \n",
    "\n",
    "For a molecule of DNA that is from an unknown species, we can analysis the frequencies of bases in the molecule and compare them to a database of DNA to attempt to determine the species. \n",
    "\n",
    "Because of the base pairing relations of the two DNA strands, cytosine and guanine will always appear in equal amounts in a double-stranded DNA molecule - this we analysis its **GC-Content**. The GC-content of a DNA string is the percentage of symbols in the string that are either 'C' or 'G'.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### FASTA format\n",
    "\n",
    "The FASTA format is a means to label strings (DNA) In this format, we introduce the label of the string using a '>' followed by the label. Then the next line is the string itself. For example,\n",
    "\n",
    "\\>Rosalind_6404\n",
    "CCTGCGGAAGATCGGCACTAGAATAGCCAGAACCGTTTCTCTGAGGCTTCCGGCCTTCCCTCCCACTAATAATTCTGAGG\n",
    "\n",
    "\\>Rosalind_5959\n",
    "CCATCGGTAGCGCATCCTTAGTCCAATTAAGTCCCTATCCAGGCGCTCCGCCGAAGGTCTATATCCATTTGTCAGCAGACACGC\n",
    "\n",
    "\\>Rosalind_0808CCACCCTCGTGGTATGGCTAGGCATTCAGGAACCGGAGAACGCTTCAGACCAGCCCGGAC\n",
    "TGGGAACCTGCGGGCAGTAGGTGGAAT\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem:\n",
    "\n",
    "Give a list of FASTA formatted DNA, compute the string with largest GC-content."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Rosalind_6489', 0.5282583621683967]\n"
     ]
    }
   ],
   "source": [
    "# Get input from file, store content in list of pairs: label, string\n",
    "file = open('question_inputs/computeGC.txt', 'r')\n",
    "\n",
    "# preprocessing\n",
    "dna_list = file.read().replace('\\n', '').split('>')\n",
    "dna_list.pop(0)\n",
    "\n",
    "# For each pair, compute GC-content of string, store label and freq.\n",
    "largest_pair = [None, -1.0]\n",
    "\n",
    "i = 0\n",
    "while (i < len(dna_list)):\n",
    "    label = dna_list[i]\n",
    "    seq = dna_list[i+1]\n",
    "    # Compute GC-content\n",
    "    count = 0\n",
    "    for char in seq:\n",
    "        if char == 'G' or char == 'C':\n",
    "            count += 1\n",
    "    if largest_pair[1] < count/len(seq):\n",
    "        largest_pair = [label, count/len(seq)]\n",
    "    \n",
    "    i += 2\n",
    "    \n",
    "print(largest_pair)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Point Mutations\n",
    "\n",
    "A **mutation** is an error that occurs during the creation or copying of a nucleci acid (in particular DNA.) Mutations come in forms, the simplest and most common is a **point mutation** which replaces one base with another at a single nucleotide. For DNA, this also changes the complementary base.\n",
    "\n",
    "For example, the pair G-C mutations to A-T.\n",
    "\n",
    "The **Hamming distance** between two strings of equal length is the number of points in which they differ. \n",
    "\n",
    "DNA from two species with a common ancestor are expected to have the similar strings (i.e. small Hamming difference.)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem: \n",
    "\n",
    "Given two strings of the same length, compute the Hamming distance between the two."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "446\n"
     ]
    }
   ],
   "source": [
    "s = \"GAGCCTACTAACGGGAT\"\n",
    "t = \"CATCGTAATGACGGCCT\"\n",
    "\n",
    "s = \"CACATAACAGCCCTGCAACCACCTAGGCCTGGAGCTACTGCTGCCAGCACTCCTGCAGTGAACTTGAAATACTTCGATGCATAATGAACTTTGTCTATTTCGGGCGCACGCAACACCGACAGCTACGATCTTCATCGAAGTTTTAGCGCCAAGATAGTTACTGCTTGTAGCCAGGTGGCGACGCTTAATAGCACGAGTGACCCTGAGGACAAATGTTGCCTCAGGTCATCAAGACGCGGCGGCTATTACTCAGTCAAACCGGCATGGGTCCCACGACCTGATACATCTCCGACTGGATGATATTATCAGATTAGCCGCAGCGTATCAAGGAACTATTCAGTAGTACTGCCTCGGGGGGTATGAACGCACGAGGCATTTAATAGTTAATGCTAACCACACTTCCAAATCATATGAAGTGGGACAGTTAGCAATGATGGCACTGGAAGAGTTCACAGTTCCTCTAAAAAGAACTAGGGTCCGGACATGGACGCTTTCGGTCGGATTTACTGCGAGTTCAAATTGGATGAAGTGGCCCACCGAATAGGTTCGAGCAACTTACTCGCAGCTATAAGGGAGTGTTATACATGTATATTTCCCCTTTTAAGCCCGTACGTGGCTCCAGATATAGACTCCTTGCAGGAAGGTGTGCACCCCTCGCTGTGCACGTGGCCGTAGATCTCGGCCCGTCCATGAGGCGATCAGTGCCAGGTGCTTTAGCCTTGTGCTGTCGGCACGTCTAGCTCCGATATGGGTCCCCTTGCCTTTTACTCACCGAAACGCGGTCGGGGGGGGTTGGTTCGGTGCCAAGGGTAAAGGGTAGCTCGAGATGTAGCGCCGTAAAACAAACGAATGGGTAATGGTATCTCCTGCAGTCTAGGTCCTATGAGCCTCTAGTGAGTATCTCGTCTCATCCACGTGCGTCAGCATAACTT\"\n",
    "t = \"CTATTAACTTATTGGCAACCAAGTACGGCTAGTGCTACAGAAACCACTAATCCTAAACTGCACCTTAAATATTTGCATGTACAACAAGTGTAGTACACATGATGCACAAGCGAGCGCGGGAGTTCGGAGCACCGATTATTATTCAGCGCGAGGAAAACCTCTTCGTTCGGTGAAGTGGCGGGGTTTTATGCCAGGTAACGGAATGACGGCAAATATACCGCCCGGTGTCCTACTTTATGCTACAACTACAATTTCGTACCGGCATGGGTCTCTCCGGCTAAGCCATTGCCCATTGGAACGTATAGCAGCGGAAGCCATATCGTATCAAAATTCTAGCCTGGGAGGCCGCCTCAGGCACAATGGACACCCGAGGTATCTAGTAGTTGGTCCTATGCACACTTCAGTACGCTTTGACACACGACAGATAGTCGAGATGGCCGAGATACGCCGCAGCTTTGCTTTAATGAGAGCTAGGTCCCGTGTTCGGATACGTTAGGCCATATTAGGAACACGTCGCACATGGATCCCCTGGAGAACCATATCGGTGTTAGCCTCTTGGCGGCCGCTGCAGAGCACTGTTGTACGTGAACATGTTCTATACCGGGCTCGTACATTACTGCGAGTCTATCTTTGGTGTAGTAGATTAGGCTTACCTAGCATTACACTCTCACCAAGATAGCTGCTACCCCAATAGTGGAAACGGACAACGGGTTTCAGTCGGGTGCCGACCGCACGGATGACATACATACGTCACCGCTCGAGGTTGACGAGTCGTAGCAAGCTCGGTAGGCCTTTTTGACGACGGAAGTGATGAGTACAGCTGTTCAATTAGAAGCTGCCCACTAAGGTGGATGTAAATCTAGCCGAGTCAGTGCAAGCCCGGCGAGGCACCGGCGAGTCTCTCGTCTAAAACAGGTGCGGCGGAGCAGCAT\"\n",
    "\n",
    "\n",
    "ham_dst = 0\n",
    "for i in range(len(s)):\n",
    "    if s[i] != t[i]:\n",
    "        ham_dst += 1\n",
    "        \n",
    "print(ham_dst)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mendelian Inheritance\n",
    "\n",
    "In 1865, Gregor Mendel described the contemporary hereditary **blending inheritance** model, which states that the inheritance must be a blend of its partent's traits. This model however is not observed in people.\n",
    "\n",
    "Mendel viewed the traits of pea plants as discrete blocks called **factors**, each of which has a distinct form called **alleles**.\n",
    "\n",
    "Mendel's **first law (law of segregation)** states that for a given factor, every organism possesses a pair of alleles for a given factor. If an individual's two allels for a given factor are the same, then it is **homozygous** for the factor; if they differ, then the individual is **heterozygous**. Thus for an offspring with two parents will receive one allele from each parent; which one is randomly decided. \n",
    "\n",
    "Further, any factor corresponds to only two possible alleles - the **dominant** and **recessive** alleles. By possessing only one copy of the dominate allele, the organism will **display the trait** of the dominant allele. That is, the only way that an organism can display a trait encoded by a recessive allele is if the individual is homozygous recessive for that factor.\n",
    "\n",
    "We encode the dominant allele of a factor using captial letters, and the recessive allele by lower case letters. As an organism may possess the recessive allele but not display it, we define an organism's **genotype** as its precise genetic makeup, and its **phenotype** as its physical manifestation of its underlying traits.\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem\n",
    "\n",
    "Given populations K,M,N of homozygous dominant, hetrozygois, and homozygous recessive factors respectively, compute the probability of any two mating organisms will produce an individual possessing a dominant allele. (An organism cannot mate with itself.) Assume at least two memebers of each population.\n",
    "\n",
    "### Solution\n",
    "\n",
    "From k,m, and n numbers in each population, we consider all events (i.e. ways to mate) along with the probability that the resulting individual will possess the dominant allele:\n",
    "\n",
    "Parents:ways that offspring can have the dominante allele.\n",
    "\n",
    "KK:4/4  \n",
    "KM:4/4  \n",
    "KN:4/4  \n",
    "\n",
    "MK:4/4  \n",
    "MM:3/4  \n",
    "MN:2/4  \n",
    "\n",
    "NK:4/4  \n",
    "NM:2/4  \n",
    "NN:0/4  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.8364182692307692\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "k=29\n",
    "m=19\n",
    "n=17\n",
    "sum_terms = k+m+n\n",
    "total_outcomes = sum_terms*(sum_terms-1)\n",
    "\n",
    "dominante_allele_outcomes = [\n",
    "    k*(k-1)*1,\n",
    "    k*m*1,\n",
    "    k*n*1,\n",
    "    \n",
    "    m*k*1,\n",
    "    m*(m-1)*0.75,\n",
    "    m*n*0.5,\n",
    "    \n",
    "    n*k*1,\n",
    "    n*m*0.5,\n",
    "    n*(n-1)*0\n",
    "    \n",
    "]\n",
    "\n",
    "print(np.sum(dominante_allele_outcomes) / total_outcomes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Translating RNA into Protein\n",
    "\n",
    "**Proteins** are chains of small molecules called **amino acids** - 20 amnio acids commonly appear in all species. Similar to nucleic acids, the **primary structure of a protein** is the order of the amnio acids.\n",
    "\n",
    "Proteins power all practical functions carried out by a cell; thus the key to understanding life lies in understanding the relationship between a chain of amino acids and the function of the protein that this chain will eventually construct. \n",
    "\n",
    "The translation of an RNA molecule (mRNA) into amino acids are how amino acids are created. Somehow, the 4 RNA bases must be translated into a language of **20 amino acids**; for every possible amino acid, we translate 3 nucleobase strings (called **codons**) into amino acids - this implies there are $4^3=64$ possible codons.  \n",
    "Two special types of codons are the **start codon (AUG)** and three **stop codons (UAA, UAG, UGA)**, these do not code for amnio acid, only indicate the translation to end.\n",
    "\n",
    "#TODO update"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem \n",
    "Given an input RNA string s corresponding to a strand of mRNA, output the protein string encoded by s. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MCHIVTRYSAHRPVHAELQFASSPTRSSDGGYKSLKAMGTTVLLRSLRSISCRTYYIIPCRSPRLEHFRLSCKISLFHARQTMLKVTVMTYPAKIAACARRAKLPFILCGGMVPSCSRVMKYGCIRYRETGKQLKCVKTALRPTNENSSIARAKCNVNVKGWLRYQLCADQGMAQLRACPMGRRGRSDMRLIGPRVLLPSGGCSKSGWYNLSLTDRPLPVSCLPMGSFKSASSNITGASPCMTIMPIIKTQNFAHCMPKLTPRITQQLKRLLLPPVFAPARPRGHSHVYQTGIAFCTPWTPEPYQSRCCHGLLAVLLIRRTFTSLQPSRFRNNIAEVSLSESIWAYCYHIEPYSLELYPEMGHASTSVSRLRPRQVAVRVRANRRISLRVQTQCGKLPTGLWVPIGKSNGDTLQPWVEVDKKFLGERTDTVETYLEICIVRPEGFSRSVELQRCDFRFWTQVSTQVTKPPVKAAVCIYNVSRRSYLSAPSNPSLPLFSQRCCNWGFSTHSEMPGDDLSPLLSVKHLHMGLRVLEAGYQDSKRHYYTTCRESSLYLRIALAVVARRKVRHHSQDDTKSLPWNNSRYKHLYVVSTTIVEYKPGAFVSRLRHNRDETPYSKNLLCLSNSSSSDGLTLQESHDIIKLIHLTQAPPSCTSHLLRDFPGTRAIEAPWMHFNKIKTNRLRLAAGIISLGGYAPASAALTRPIHIISGVTRWLPPLTYPTTSGTSNVALPHAAWVGCDTSDLHFVKHPTLRRITVKGLTRVTSLLRFVNNSPVELEDVTFVFTRPFEVQRLLLSARQCGPSRKSDKPYSQFNELPVEYNISRRSTRPDHSNPLRAQRERQVLEPDIDLCSHGMLLFRHPPGRLCLSIRSTNPRAPRKSGKAHDVDHRPHGLRCRVLRFEASSAPGSVWSAFCNNLPLGSLLACPVGAMLRRAVREAKLTVTGVTLTNVIANETWASGPPLGRRRLSGIALTGSLHCPRAVSPYCLLRSLICCEVLGYGAGCLLRLISSQSVLRVMLEARRTTLGIGSIIATKFTCLLTQVWQEAANHKCHGARIRMSLQSFGTGRTIASNNAPRRSSVELGHQPCHLRLAKFAVKDRPIAIYRPINANEWGYGIATKNTSFYHRRILAATGSFNTKSSGKGGVKSTDLELYRSSNKMSSVTRRSMLSTLALSKSLRRSRRIVQLAPRRPRHWNFSDVYSPLTIAGAYHSDKQTLALTSRSAYLVGRSVTLLTKRKDYDSLLEFAFLFDTALSALSSIPLSVWVRPKNLTRFSILNISYLLPCVNDVTAALSRTTNFGCILKSAIRYYRPDKHSPVNVAVYPDAERIRYRVNAERTLVTLRVGALCLYPTGFVDNAADVALYAYGVYVGGLEPRPTLYKIVSGQQAAIQNRAEIQTTARYVTVPSDVNDLGEATPALTKVNISVISVPLSRLSGSTGMKAPTFEAPTTGSESECSIRCPLVMKHGPWLENGFRSVASRAVAKSITFEEIGYKARTKANTFCRPSSCRLNFRSLLPWGFREPNSVNGSGVAARVLICSLKPPLNREEEKIRAVRGLLWSVYAKKARACRVGHSGSNLPMSDPDPRRLLVPDSRRDYCNYEDSKGIIPTTNYCARRVPPYRKTGPIARGSQPAVKDSPSMRPAATRSFKRHARAKQHAMVVLSKLLLDGGHCCTIPFRSYTKLSMAVRPGQVLSRAKMLFKILSFHEVLTTRLLHTFGLHSFTMEAYTVSLYSYAARVCSHDPLVGLYYGTKDTRIGPVRNLRLSLRPIRDILSRGHLVPTVPVTSLGRNKTGVNIRTFSSAPLYDGIEFGPHLSALACSSSHYVRPVGDIHRINAQYLIQRPQRQLPFNSETPVSQNELLTVNLYQYFTCEPIDHCPRTSDTSTLEDLMPLVPRPLPLSATMYTDALYTHVSKTLLSYSYDQHKPPTINALKGEGAPIWGPLSIDRVTRLASTVEILRLSDYESTFYPPTPDGDSYLTKLYKSMTPAQCVRLMLEPKRTEKTRGGNSGKLQSDAQSCADLQRTLSTKLRWRSEAAEDALTSPATDDIVFCPAHHLRIHPFSRYMQSLTSPKATRELKMGILPIPLPPYSRHCASYISISIIGHGKSLCPCGLATVAYPHRDDITNQVSSSVKNLVLQSMCPRILIQHLLSIVRVLVVNSTLCHWIVRYVYPKCDEYVSGDNIVACQSRLLLRVYSDELTASLAKMSSRFQTDLLGRVRGRLPAPELTVRAAGSPNEESTKLTSEHLVSPLLSCPQECFNRQNIHVLISFSQAYADSGVVQCQNGDSLHIFHSNSRRIVLGVMSFMLSAKVGHRNYGGSDDNADSHTHRGSRGDLSADFLVKSRGRVWRSVWYLLKNAISGLRCPSWIWSNTIAYQASRTIMSIGRLRQINQGQISPRESGGQTTSVQGVKNETGSICSKGFFIVWSSVEAPTFHDDMQARSSIGQAIIRCASELDARVRDVETVGSGADGLPSESSSTCSSFRKLMVGTPPPAVHETKVRKTSAMRARTDCLAQRGEPVKVLVSSPLGRNWTALISRGRYERHPTEPTKSSQLDETTRAMRQEIRKPSSVSCYQCSKEGHKHRACVSRLAWFINWVHNKYTMAPSLLHCGGRGVSRVALNTFKIECFLSPVLRARYASAERSLQHDLRVLSTPVLSNLHGDTEESCGIYSVALQGCLLYHSPKRQYSYPQSPTRVGRNLSKRVCKAQVRSARWIALQTNKYFCDDGSPSGLVFTARAAFFTEAFPHLALIQLKNTLLNVAHNSAVPRDDSDCHLSELSYPTARMSLPSRREPQDKVSWLADWKTPRIQTPYRQLVTSSYHKLPSVESAISLDCAHSLLYVGLKVAVDSSGWCVSYWHRHPDTPGTRTTDAARLGNTMSDTTNAKVWHPRKGLCSTNPLIVGPTTVLPSSVSRSSTRLHWEKMLLNWQSRAGDCRTTNVGPCRCAVNKGRIGIVLRTCHESAEAHSAQRITTRSSSRRPLGLTVNYHLSKSSLAWQYVAGEYIYFTEGLPLHRNRNHCYAKMSVEFITMLTECTSGENGKISTRAVFRTEYSYSSIGSNGKSEASGTRCGACRAKRAGLNASNVTDCLGPWAAALSNVVRFGVGLKVLLLSAWGVHAAKSQACPYLRQAGETNSAIWVEEVVVRPNESQIVGGCSNIWREVSLKSLRIYRYKSLPNTDRRLSDRMNADTYTSTPTLKYSGSTLARSIQGCVKKMQSTTQLRLRDPTLHSYLLQLELVLATPLPRM\n"
     ]
    }
   ],
   "source": [
    "# Create dict where k-v pair corresponds to RNA condon-amino acids for the purpose of protein creation.\n",
    "dict = {}\n",
    "with open('question_inputs/rna_codon_table.txt', 'r') as f:\n",
    "    for line in f:\n",
    "        k,v = line.split()\n",
    "        dict[k] = v\n",
    "        \n",
    "# input string\n",
    "s = 'AUGUGCCAUAUUGUUACCAGAUAUUCUGCACACCGCCCGGUACAUGCGGAACUUCAAUUUGCAAGCUCACCUACGCGCUCUUCCGACGGGGGGUACAAAAGCUUGAAGGCCAUGGGUACUACCGUUCUUCUCCGGUCGUUGCGGUCCAUAUCUUGCCGUACUUAUUAUAUUAUACCCUGCCGCUCGCCGAGGCUCGAGCACUUUCGCCUCAGCUGCAAGAUCUCGCUGUUCCACGCGCGUCAGACAAUGCUUAAGGUUACAGUAAUGACGUACCCCGCAAAAAUUGCGGCGUGCGCAAGGAGGGCUAAACUACCCUUCAUACUGUGCGGUGGUAUGGUGCCUAGCUGUUCCCGUGUCAUGAAAUACGGGUGUAUCCGCUACCGUGAGACGGGAAAACAACUCAAGUGCGUUAAGACGGCCCUACGUCCUACUAAUGAGAAUAGUUCUAUUGCCCGUGCGAAGUGCAAUGUCAAUGUGAAAGGUUGGCUACGGUAUCAGUUAUGCGCCGAUCAAGGUAUGGCGCAACUACGCGCCUGCCCUAUGGGUCGAAGGGGUCGAUCGGACAUGAGGUUGAUAGGACCUCGCGUGCUACUACCAUCGGGCGGCUGCAGCAAAAGUGGCUGGUACAAUCUUAGUCUAACUGACAGACCGCUGCCGGUCUCAUGUUUACCCAUGGGCUCGUUCAAGUCUGCAUCGUCGAAUAUUACAGGGGCGAGCCCCUGUAUGACUAUAAUGCCAAUUAUCAAAACACAGAAUUUUGCUCAUUGUAUGCCAAAAUUGACCCCGCGGAUAACUCAGCAGUUGAAACGACUUCUGCUGCCUCCGGUGUUUGCUCCCGCUAGGCCACGAGGCCAUUCACACGUGUAUCAGACCGGGAUUGCUUUCUGCACACCGUGGACACCUGAGCCCUACCAAUCGCGAUGUUGCCAUGGGUUGCUUGCAGUUCUAUUGAUAAGGCGCACAUUUACUUCCUUGCAACCGAGCCGCUUCCGUAAUAACAUAGCCGAGGUAUCACUCUCGGAGAGCAUUUGGGCAUAUUGCUACCACAUCGAGCCGUACUCACUUGAGCUAUACCCUGAAAUGGGUCAUGCAAGUACCUCCGUAAGUCGGCUCCGUCCCCGACAAGUAGCGGUGCGAGUAAGAGCAAACAGACGCAUAAGUCUCAGAGUCCAAACGCAAUGUGGGAAGUUGCCCACUGGUUUGUGGGUACCCAUCGGUAAAUCUAACGGAGACACUCUGCAACCGUGGGUUGAAGUAGAUAAAAAAUUCCUGGGUGAACGUACGGACACUGUAGAGACUUAUUUAGAAAUUUGCAUCGUCAGGCCGGAGGGCUUUUCUAGAUCUGUAGAGUUACAGAGAUGCGAUUUCAGGUUCUGGACGCAAGUUUCCACUCAAGUCACAAAGCCACCAGUGAAAGCUGCCGUAUGCAUCUACAAUGUCUCCCGGCGUUCAUACUUAUCGGCGCCUAGCAACCCGUCCCUCCCCCUUUUUAGCCAACGAUGCUGUAACUGGGGUUUCUCCACACACUCAGAAAUGCCGGGGGACGAUCUUUCUCCUUUGCUGUCGGUAAAACAUUUACACAUGGGGCUACGGGUACUGGAAGCAGGGUACCAAGACUCCAAACGUCACUAUUAUACUACAUGCCGGGAGUCUUCACUUUAUUUGCGAAUUGCGCUUGCGGUGGUAGCGCGGCGCAAGGUCCGGCAUCACAGCCAAGACGAUACUAAGAGUCUACCUUGGAACAACAGUCGAUAUAAACAUCUCUAUGUCGUAAGUACUACAAUUGUUGAGUACAAGCCGGGCGCCUUUGUAUCGAGACUCCGGCACAACCGCGACGAAACUCCAUACUCCAAGAACUUGCUCUGUCUAAGCAAUAGUUCAUCAUCGGAUGGAUUGACCCUACAGGAGAGCCACGACAUUAUAAAACUGAUCCACUUGACGCAGGCUCCUCCCUCUUGCACUAGUCACCUACUAAGAGACUUUCCAGGGACACGGGCGAUAGAAGCACCAUGGAUGCAUUUCAAUAAGAUCAAGACCAACAGGCUGAGGCUCGCGGCAGGCAUCAUUUCCCUUGGCGGCUACGCACCUGCAAGCGCGGCGCUUACACGUCCGAUUCAUAUUAUUUCGGGCGUGACGCGUUGGCUGCCCCCCCUAACAUAUCCAACCACAAGUGGAACAUCCAAUGUGGCACUCCCACACGCUGCGUGGGUGGGAUGUGAUACUUCAGACCUACAUUUUGUUAAACAUCCAACUCUACGUCGAAUUACGGUUAAAGGGCUCACAAGAGUGACUUCACUAUUAAGGUUUGUAAAUAAUUCUCCCGUGGAACUGGAAGACGUUACCUUUGUGUUUACGCGUCCAUUCGAAGUCCAGCGACUCUUGCUAAGCGCACGUCAAUGCGGCCCGUCACGUAAAUCUGACAAGCCGUACUCACAGUUCAAUGAGCUACCAGUGGAAUACAACAUAUCCCGGAGGAGCACACGCCCUGAUCACUCCAAUCCUUUACGGGCACAGCGGGAACGUCAAGUACUUGAACCGGACAUUGAUCUUUGUAGUCAUGGCAUGUUGUUAUUUCGUCAUCCGCCAGGUCGUUUAUGCCUGAGCAUUCGAUCGACCAAUCCGCGCGCUCCCAGGAAGUCAGGAAAAGCCCACGAUGUCGAUCAUCGUCCUCACGGACUCAGAUGCCGCGUGCUGAGAUUCGAAGCGAGUUCCGCUCCAGGUAGUGUGUGGUCCGCCUUCUGCAAUAAUCUGCCCUUAGGGUCUUUGCUCGCAUGCCCGGUGGGUGCUAUGUUAAGAAGGGCGGUUCGAGAAGCGAAGCUAACGGUGACAGGUGUUACUUUGACAAACGUGAUAGCAAACGAGACGUGGGCCUCUGGACCCCCGCUUGGUAGGCGACGUCUCUCCGGCAUUGCGCUUACGGGUUCCCUUCACUGCCCUCGUGCAGUCUCACCCUAUUGCCUUCUGAGGUCUCUCAUAUGUUGUGAAGUACUAGGUUAUGGGGCCGGCUGCUUACUUCGUUUGAUCAGCAGCCAGUCGGUGUUGAGAGUUAUGCUAGAGGCGCGACGCACCACACUCGGGAUCGGCAGUAUUAUAGCAACAAAAUUCACGUGCUUGUUAACCCAAGUGUGGCAAGAAGCAGCCAACCACAAGUGCCACGGCGCUAGGAUUAGGAUGUCUUUACAGUCUUUUGGAACCGGUCGUACGAUAGCGAGUAAUAACGCCCCCCGAAGAUCGUCGGUUGAGUUGGGUCACCAGCCAUGUCAUCUCCGGCUCGCUAAAUUUGCUGUAAAAGAUCGUCCGAUAGCCAUAUAUAGGCCAAUCAACGCCAAUGAAUGGGGCUACGGUAUUGCAACAAAGAAUACUAGUUUCUACCAUCGUAGGAUACUGGCGGCCACCGGGUCCUUCAAUACUAAAAGUUCGGGGAAAGGGGGGGUUAAGUCUACUGACCUAGAACUAUAUCGGAGCUCGAAUAAGAUGUCGUCGGUCACGCGGCGCAGCAUGCUAUCUACGUUAGCCCUAUCGAAAUCGCUGCGAAGGUCGCGUCGAAUCGUUCAGCUCGCGCCUCGUCGACCCCGGCAUUGGAACUUUUCCGAUGUAUAUUCGCCGCUGACAAUUGCGGGCGCGUAUCAUAGCGACAAGCAAACUCUAGCACUGACUUCUCGGUCGGCAUACUUGGUAGGUAGAUCAGUAACGCUCCUUACUAAACGUAAAGAUUAUGAUUCACUAUUAGAGUUUGCGUUUCUAUUUGAUACUGCGCUUUCCGCUCUAUCAUCUAUCCCCCUCUCCGUUUGGGUUCGUCCAAAAAACCUCACAAGAUUCUCUAUAUUGAAUAUCUCAUAUCUGCUUCCGUGUGUGAAUGACGUCACAGCAGCACUGUCCCGCACAACAAAUUUCGGUUGUAUCCUCAAAUCGGCGAUCCGCUAUUACCGGCCGGAUAAACACAGCCCUGUUAAUGUAGCCGUUUAUCCAGACGCAGAGCGGAUCAGAUAUCGAGUAAACGCGGAACGUACCCUAGUAACCCUUCGUGUAGGUGCUUUAUGCCUUUACCCUACUGGCUUUGUCGAUAAUGCAGCUGAUGUUGCUCUCUAUGCUUACGGGGUCUACGUCGGAGGCUUAGAACCUCGUCCUACACUUUAUAAAAUAGUAUCCGGGCAACAAGCUGCUAUCCAGAACAGGGCCGAGAUACAGACAACCGCUAGGUACGUAACAGUACCAAGCGACGUUAACGAUCUUGGAGAAGCUACACCGGCCUUAACUAAGGUAAAUAUCUCAGUGAUAUCUGUUCCACUGUCACGUUUGUCGGGCUCCACCGGAAUGAAAGCCCCUACAUUUGAAGCACCUACAACUGGAUCCGAAAGCGAGUGCUCUAUACGUUGCCCUUUGGUUAUGAAACAUGGGCCAUGGCUUGAGAAUGGUUUCCGUAGUGUCGCCUCUAGGGCCGUCGCAAAAUCCAUUACGUUUGAAGAGAUCGGGUACAAGGCGAGGACCAAAGCCAAUACGUUCUGUCGGCCAAGCUCCUGUAGACUCAAUUUUCGUUCGCUGCUACCUUGGGGGUUUAGAGAACCGAAUUCGGUCAAUGGGUCGGGUGUAGCAGCGCGAGUGCUAAUAUGUAGUCUGAAGCCGCCGCUCAAUCGGGAAGAAGAAAAGAUUCGGGCAGUAAGGGGACUGUUAUGGUCGGUAUAUGCCAAAAAGGCCCGAGCAUGCCGUGUCGGGCAUUCAGGCAGCAAUCUGCCAAUGUCGGACCCUGACCCUAGAAGGUUGCUAGUCCCGGAUAGUCGCCGGGAUUAUUGUAAUUACGAAGAUUCAAAAGGCAUAAUCCCGACCACGAAUUAUUGCGCCCGCAGAGUACCACCCUACCGCAAAACAGGGCCAAUAGCGCGUGGAAGCCAACCGGCAGUGAAAGAUUCCCCUUCAAUGCGACCAGCCGCAACUCGCUCAUUCAAAAGACACGCCCGUGCCAAACAGCACGCUAUGGUAGUACUCUCUAAACUAUUACUUGAUGGUGGGCACUGCUGCACGAUUCCCUUCAGGUCGUAUACAAAAUUGUCGAUGGCUGUUAGGCCGGGACAAGUCCUUUCCAGGGCAAAGAUGUUGUUUAAGAUCCUAUCUUUUCAUGAGGUACUGACUACCCGGCUGCUGCACACUUUUGGUUUGCAUUCGUUUACAAUGGAAGCCUACACGGUGUCCCUUUAUUCUUAUGCGGCGCGUGUGUGCUCCCAUGACCCUCUAGUGGGCUUGUACUACGGCACUAAAGAUACCCGUAUCGGGCCUGUCCGUAACUUGCGGUUAAGUCUGAGGCCAAUACGAGAUAUCCUUUCGCGUGGACACCUGGUCCCCACAGUUCCCGUCACAAGCCUGGGGAGGAAUAAAACCGGGGUAAAUAUUCGCACGUUUUCAUCGGCGCCCUUGUAUGACGGAAUUGAGUUUGGCCCACAUCUCAGCGCCCUCGCUUGUUCGUCAAGCCACUACGUUAGACCCGUGGGGGAUAUCCAUCGUAUAAAUGCUCAAUAUUUGAUUCAGCGUCCUCAGCGCCAGUUGCCGUUCAAUUCAGAAACCCCGGUAUCGCAGAACGAACUCCUCACCGUCAAUCUGUAUCAAUAUUUCACUUGCGAGCCCAUAGACCAUUGUCCGCGCACUUCCGAUACCUCGACGCUUGAGGACCUAAUGCCACUAGUGCCUCGACCUCUUCCGCUAUCGGCUACCAUGUACACGGACGCGCUCUAUACCCAUGUGUCCAAAACUCUACUAAGCUAUAGUUACGACCAACACAAACCUCCGACCAUAAAUGCGCUAAAGGGCGAGGGCGCGCCAAUAUGGGGGCCCCUGUCAAUCGAUAGGGUUACUCGACUGGCGUCAACUGUGGAGAUUCUUAGAUUAAGUGACUAUGAAUCAACCUUUUACCCGCCUACGCCGGACGGCGAUUCAUACUUGACGAAGUUAUACAAGUCGAUGACACCCGCCCAGUGCGUGAGACUCAUGCUUGAGCCAAAGCGUACCGAAAAAACUCGGGGGGGGAAUUCCGGUAAGUUGCAGAGCGAUGCACAAUCAUGUGCGGACCUUCAAAGGACUCUUUCGACCAAAUUAAGAUGGCGCAGCGAGGCGGCAGAGGAUGCACUUACUAGUCCUGCGACUGACGAUAUCGUUUUCUGCCCAGCCCAUCACCUUCGAAUACAUCCCUUCUCCCGCUAUAUGCAGUCCCUUACCUCUCCGAAGGCUACCAGAGAACUCAAGAUGGGGAUACUACCAAUACCUCUCCCGCCUUACUCCCGCCACUGUGCGUCUUACAUAUCGAUUUCUAUAAUAGGGCAUGGGAAAUCGCUAUGCCCAUGCGGACUUGCCACCGUGGCCUACCCUCACCGUGACGAUAUCACUAAUCAAGUCUCAAGCUCAGUUAAAAACUUGGUUUUACAAUCAAUGUGCCCUCGUAUUCUAAUUCAGCACCUUCUGUCAAUUGUUCGUGUCCUUGUGGUUAAUAGCACUUUAUGUCAUUGGAUAGUACGAUAUGUCUACCCCAAAUGCGACGAAUAUGUGUCAGGUGACAACAUCGUGGCUUGCCAAUCCCGACUCCUCCUGCGGGUCUACAGCGACGAGCUCACAGCUUCCCUAGCAAAAAUGAGUUCCCGCUUUCAAACUGAUUUGCUGGGCCGGGUGAGAGGGAGACUGCCGGCACCAGAAUUAACUGUACGGGCUGCGGGGAGUCCUAACGAAGAAUCAACCAAACUUACCAGCGAGCACCUUGUUUCCCCCUUGCUAUCAUGCCCGCAAGAAUGUUUUAACAGGCAGAACAUACAUGUACUCAUAUCCUUUAGCCAGGCGUAUGCUGACUCAGGAGUCGUCCAAUGCCAGAAUGGUGAUAGUUUACAUAUCUUCCACUCAAACAGCCGCAGAAUCGUCCUGGGCGUUAUGUCAUUCAUGCUAUCGGCAAAGGUUGGGCACCGGAAUUAUGGUGGAAGUGAUGACAAUGCAGACAGCCACACACAUCGCGGUAGCCGUGGGGAUUUAUCGGCUGACUUCUUAGUUAAGUCGCGCGGUCGUGUAUGGAGAAGUGUAUGGUAUCUCCUGAAGAAUGCAAUCUCCGGUUUAAGAUGUCCCUCCUGGAUAUGGAGUAAUACCAUCGCUUAUCAGGCGAGUAGGACGAUAAUGUCAAUUGGGCGUCUCCGACAGAUUAACCAGGGUCAGAUCUCUCCACGUGAGAGUGGGGGUCAAACGACCAGUGUGCAGGGCGUUAAAAAUGAAACAGGCUCCAUAUGCUCCAAAGGAUUCUUUAUCGUAUGGUCGAGUGUUGAGGCUCCAACGUUCCACGAUGAUAUGCAAGCUAGGUCUUCUAUUGGGCAAGCAAUUAUACGCUGUGCCAGUGAGCUGGAUGCGCGCGUACGAGAUGUUGAAACGGUGGGAUCCGGGGCAGACGGGCUCCCAUCGGAGAGCUCUUCAACAUGUUCAAGCUUUCGGAAGCUGAUGGUUGGUACUCCGCCACCCGCGGUGCACGAGACCAAAGUAAGGAAAACCUCUGCGAUGCGUGCCAGGACAGAUUGCUUAGCACAGCGCGGGGAGCCAGUGAAGGUCCUGGUAUCGUCUCCAUUAGGUAGAAACUGGACUGCUCUAAUAUCUAGAGGGCGCUAUGAGAGGCAUCCAACUGAGCCUACGAAAUCUAGCCAGCUUGAUGAGACGACGAGAGCUAUGCGACAAGAGAUACGCAAGCCCAGUUCCGUUAGUUGCUAUCAAUGUUCGAAAGAGGGACACAAGCAUAGAGCGUGCGUUUCACGCCUGGCUUGGUUUAUAAAUUGGGUUCACAACAAGUACACGAUGGCACCUAGCCUCUUGCAUUGCGGAGGACGUGGCGUUAGCCGCGUCGCACUAAACACUUUUAAAAUCGAAUGUUUCUUAAGCCCAGUACUGCGAGCACGUUAUGCCAGUGCAGAACGUUCCCUACAGCAUGAUUUGAGAGUCCUCAGUACUCCUGUCCUGUCAAAUUUGCACGGGGACACCGAAGAAAGUUGCGGGAUUUACAGCGUGGCGUUACAGGGUUGUCUUCUCUACCACUCACCCAAGAGACAGUAUUCCUAUCCGCAAUCACCUACGCGUGUAGGCAGAAACCUCAGCAAACGCGUGUGCAAGGCGCAAGUUCGUUCCGCACGAUGGAUUGCCUUGCAAACGAAUAAAUACUUCUGCGAUGACGGGAGCCCCUCAGGCCUUGUCUUCACGGCACGAGCAGCCUUCUUUACCGAGGCGUUUCCCCAUCUAGCCCUCAUCCAGCUUAAGAAUACAUUGCUUAAUGUUGCACAUAACAGCGCGGUACCCCGAGAUGAUUCCGACUGUCAUCUCUCGGAAUUGUCGUACCCAACGGCCCGGAUGUCGUUGCCCUCAAGAAGGGAGCCGCAAGACAAGGUCAGUUGGUUAGCAGAUUGGAAGACACCUCGCAUUCAAACACCUUAUCGUCAACUCGUUACGUCUAGUUACCAUAAAUUACCCUCCGUUGAGUCGGCAAUAAGUCUCGACUGUGCCCAUUCUCUGCUUUAUGUGGGUCUUAAGGUUGCGGUCGAUUCGUCGGGAUGGUGCGUAUCUUAUUGGCACAGACACCCUGACACGCCAGGAACACGAACUACAGAUGCUGCUAGGCUAGGGAACACUAUGAGUGACACAACCAACGCGAAAGUUUGGCAUCCGCGAAAAGGGCUUUGCUCUACGAACCCACUUAUUGUAGGCCCGACAACGGUGCUCCCUUCGUCCGUCUCUAGAUCCUCGACUCGUUUGCAUUGGGAGAAGAUGUUACUGAACUGGCAAUCCCGUGCAGGGGAUUGUCGGACGACAAACGUUGGUCCAUGUAGAUGCGCCGUUAACAAGGGACGUAUCGGGAUAGUUCUGCGGACAUGCCACGAAUCCGCCGAAGCACAUUCUGCUCAACGAAUUACUACAAGAUCCUCCUCUCGGCGUCCGCUAGGUCUCACCGUUAACUACCACCUUAGCAAAAGCAGCCUGGCCUGGCAGUAUGUGGCAGGGGAAUAUAUCUACUUUACCGAGGGUUUACCCUUGCAUAGAAAUCGGAAUCAUUGCUAUGCUAAAAUGAGCGUGGAGUUUAUCACGAUGUUGACUGAGUGCACCAGUGGCGAGAACGGCAAGAUCUCUACUCGUGCAGUCUUCCGGACAGAGUAUUCUUACUCAUCUAUUGGCUCGAACGGUAAGAGUGAAGCUAGCGGUACCAGGUGUGGGGCUUGUCGAGCCAAGAGGGCCGGUCUCAAUGCCUCAAACGUGACUGAUUGCCUAGGUCCUUGGGCAGCGGCUUUAUCGAAUGUGGUGCGAUUUGGCGUGGGACUGAAAGUUUUAUUGCUGUCUGCAUGGGGCGUCCACGCGGCUAAGUCACAGGCGUGUCCUUAUCUGAGACAGGCUGGAGAGACUAACUCGGCCAUAUGGGUUGAGGAGGUCGUUGUGAGGCCGAAUGAAAGUCAAAUCGUCGGUGGUUGCUCGAACAUUUGGAGAGAGGUAAGUUUGAAAAGUCUUCGCAUCUACAGGUACAAGUCUUUGCCAAACACCGAUCGGAGACUAUCCGAUCGCAUGAACGCCGAUACUUACACUUCCACGCCAACCCUUAAAUAUUCGGGGAGUACCCUUGCACGAUCAAUCCAAGGUUGUGUGAAGAAAAUGCAGUCGACUACCCAACUCCGUCUCCGCGAUCCGACUUUGCACAGUUACCUGCUGCAGCUUGAACUAGUUCUUGCGACUCCUCUCCCGCGCAUGUAA'\n",
    "output = ''\n",
    "i = 0\n",
    "while i < len(s):\n",
    "    codon = s[i:i+3]\n",
    "    amino_acid = dict[codon]\n",
    "    if amino_acid == 'Stop':\n",
    "        break\n",
    "    else:\n",
    "        output += amino_acid\n",
    "    i += 3\n",
    "print(output)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Finding a Motif in DNA\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If the same interval of DNA in the genomes of two different organisms is found together, it is suggestive that the interval has the same function in both organisms. \n",
    "\n",
    "A **motif** is such a commonly shared interval of DNA. A common task is to search for an organism's genome for a known motif. \n",
    "\n",
    "Genomes are riddled with intervals of DNA that occur multiple times (perhaps with only slight modifications) called **repeats**. The occur too often to be simply random chance. Thus the language of DNA is quite powerful.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem\n",
    "\n",
    "Given two string s and t, report the starting index of all occurances of t in s, using 1-based indexing. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[3, 10, 17, 56, 68, 126, 133, 215, 222, 233, 264, 272, 331, 379, 394, 401, 408, 432, 470, 486, 493, 501, 508, 637, 681, 688, 695, 712, 762, 795, 803, 810, 825, 847, 874, 881, 891, 898, 905, 920, 927, 952]\n"
     ]
    }
   ],
   "source": [
    "s = \"ATGGCACGAGGCACGAGGCACGAGGGCGGCACGATTCAGGCACGAAGGCACGACTGGCACGAGGCCGGGCACGAGGCACGAAGGCACGATGGCACGAAGGCACGACGGCACGAAAGAGGCACGAAGGCACGAGGCACGAGGCACGACCGAGGCACGAATCGGCACGATCCCTTTTAGGCACGATGAGCTCGGCACGATCCGCCAAGGCACGACTGGCACGAGGCACGAGGACGGCACGAGGCACGATTCCGGCACGATACCTGGGCACGAGGGCACGAGGTACACGACGGCACGAAACGGCACGACTACCGGCACGAAGGCACGACTACGGGCACGAGGCACGATGGTGGCACGACTGCGGCACGACGGCGGCACGACGGCACGAGGCACGAAGGCACGAGGCACGAGGCACGAGGCACGAGCCTTTCTAAGGCACGAGGCACGAAGCGTCAATTAGGCACGACAGCGCGGCACGAGGGCACGATGGCACGAGGCACGAGGGCACGAGGCACGAGGCCCACGGCACGAAATTGGGCACGATCGGCACGAGAGCAGGGCACGAGTCGAGGCACGATCGGGCACGAGCAGGCACGAGACGGCACGATTCTGGCACGAGTGGCACGAAGGCACGATCTGGGCACGAGGCACGATAAGGCACGATGGCACGAAGGGGCACGACGGGCACGAGGCACGAGGCACGAGGCACGAGATGGCACGAGGGCACGAAAGGCACGATCCCGTCACGGCACGACAGGCACGATGGCACGAGGCACGATCTGAGTGGCACGATCGGCGGCACGAGGGCACGAGGCACGAGGCACGACGGCACGAGGCACGACTGATGTGGGCACGAGGCACGATAGTCGGCACGACGGCACGAGGCACGAGGGGGCACGAGGCACGAGGCACGAGGCACGACGGCACGAGGCACGAGGCACGAGCGGCACGAACGGCACGAGGCACGA\"\n",
    "t = \"GGCACGAGG\"\n",
    "\n",
    "index_loc = []\n",
    "\n",
    "for i in range(len(s) - len(t)):\n",
    "    if s[i:i+len(t)] == t:\n",
    "        index_loc.append(i+1)\n",
    "        \n",
    "print(index_loc)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Consensus and Profile\n",
    "\n",
    "We can count the point mutations between two strings, but for multiple strings (of equal length) we can do more. For several homologous strands, we can find an average-case strand to represent the most likely common ancestor of the given strands. \n",
    "\n",
    "For each index of the the strings, by taking the largest number A,C,G, and T bases as a representative of all the strings forms the **concensus string**. The **profile** of the strings (all of length n) is the 4xn matrix where entry $1,j$ denotes the count of A in all strings at position j, $2,j$ the count of C in all string at position j, and so on.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem \n",
    "\n",
    "Given a collection of DNA strings of equal length, all in FASTA format, output the concnsus string along with the resulting profile."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Helper function: read FASTAf file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_FASTA_file(file):\n",
    "    # preprocessing\n",
    "    dna_list = file.readlines()\n",
    "\n",
    "    dict = {}\n",
    "\n",
    "    curr_key = ''\n",
    "    curr_string = ''\n",
    "\n",
    "    for item in dna_list:\n",
    "        # Line contains label\n",
    "        if item[0] == '>':\n",
    "            # store current_key with current_string\n",
    "            if len(curr_string) != 0:\n",
    "                dict[curr_key.replace('\\n','').replace('>','')] = curr_string.replace('\\n','')\n",
    "                curr_key = ''\n",
    "                curr_string = ''\n",
    "\n",
    "            curr_key = item\n",
    "        # line contains piece of string. \n",
    "        else:\n",
    "            curr_string = curr_string + item\n",
    "\n",
    "    # Add last pair to dict.\n",
    "    dict[curr_key.replace('\\n', '').replace('>','')] = curr_string.replace('\\n', '')\n",
    "    file.close()\n",
    "    return dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'>Rosalind_1': 'ATCCAGCT', '>Rosalind_2': 'GGGCAACT', '>Rosalind_3': 'ATGGATCT', '>Rosalind_4': 'AAGCAACC', '>Rosalind_5': 'TTGGAACT', '>Rosalind_6': 'ATGCCATT', '>Rosalind_7': 'ATGGCACT'}\n"
     ]
    }
   ],
   "source": [
    "print(read_FASTA_file(open('question_inputs/consensus.txt', 'r')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "# Get contents of FASTA file\n",
    "dict = read_FASTA_file(open('question_inputs/consensus_input.txt', 'r'))\n",
    "\n",
    "output_file = open('question_outputs/consensus_output.txt', 'w')\n",
    "\n",
    "# create profile matrix\n",
    "\"\"\"\n",
    "row 0 -> A\n",
    "row 1 -> C\n",
    "row 2 -> G\n",
    "row 3 -> T\n",
    "\"\"\"\n",
    "list_strings = list(dict.values())\n",
    "P = np.zeros(shape=(4, len(list_strings[0])))\n",
    "P = P.astype(int)\n",
    "\n",
    "# Add values to profile matrix.\n",
    "for string in list_strings:\n",
    "    for i,char in enumerate(string):\n",
    "        if char == 'A':\n",
    "            P[0][i] += 1\n",
    "        elif char == 'C':\n",
    "            P[1][i] += 1\n",
    "        elif char == 'G':\n",
    "            P[2][i] += 1\n",
    "        else:\n",
    "            P[3][i] += 1\n",
    "\n",
    "# Output results, print to file. \n",
    "# output must be formatted carefully!\n",
    "ave_string = \"\"\n",
    "for i in  range(len(P[0])):\n",
    "    idx = np.argmax(P[:,i])\n",
    "    if idx == 0:\n",
    "        ave_string += \"A\"\n",
    "    elif idx == 1:\n",
    "        ave_string += \"C\"\n",
    "    elif idx == 2:\n",
    "        ave_string += \"G\"\n",
    "    else:\n",
    "        ave_string += \"T\"\n",
    "\n",
    "def get_line(P,i):\n",
    "    send = \"\"\n",
    "    for j in range(len(P[i])):\n",
    "        send += str(P[i][j])\n",
    "        if j != len(P[i])-1:\n",
    "            send += \" \"\n",
    "    return send\n",
    "        \n",
    "output_file.write(ave_string +'\\n')\n",
    "output_file.write('A: ' + get_line(P, 0) +'\\n')\n",
    "output_file.write('C: ' + get_line(P, 1) +'\\n')\n",
    "output_file.write('G: ' + get_line(P, 2) +'\\n')\n",
    "output_file.write('T: ' + get_line(P, 3))\n",
    "\n",
    "output_file.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mortal Fibonacci Rabbits\n",
    "\n",
    "Recall the Fiboncacci sequence for the number of rabbits after n months from earlier. We expand on this sequence by specifying that rabbits will 'die out' after living for m months.\n",
    "\n",
    "We assume that we start with one pair of rabbits; it takes a rabbit one month before being able to reproduce in which a single pair will produce a single pair of offsprings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "134779554541299204933\n"
     ]
    }
   ],
   "source": [
    "import math\n",
    "import sys\n",
    "np.set_printoptions(threshold=sys.maxsize)\n",
    "\n",
    "n = 98\n",
    "m = 19\n",
    "\n",
    "#Note: Values of matrix will become large; make sure data type of matrix can handle it!\n",
    "dp_matrix = np.full((n), -math.inf, dtype=object)\n",
    "dp_matrix[0] = 1\n",
    "dp_matrix[1] = 1\n",
    "\n",
    "def helper(i):\n",
    "    # Set value\n",
    "    if dp_matrix[i] == -math.inf:\n",
    "        # First m values are given by $F_n = F_{n-1} + F_{n-2}$\n",
    "        if i < m:\n",
    "            dp_matrix[i] = helper(i-1) + helper(i-2) \n",
    "        # The remainder is given by $F_n = F_{n-2} + F_{n-3} +  ... + F_{n-m-1}$\n",
    "        else:\n",
    "            dp_matrix[i] = 0\n",
    "            for j in range(m-1):\n",
    "                dp_matrix[i] += helper(i-2-j)\n",
    "    return dp_matrix[i]\n",
    "\n",
    "print((helper(n-1)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Overlap Graphs\n",
    "\n",
    "**Overlap Graph**: Given a collection of strings and a positive integer k, the overlap graph for the strings is a directed graph $O_k$ where each string is a vertex and a directed edge $(s,t)$ exists if there is a $k$ length suffix of $s$ that matches a $k$ length prefix of $t$. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Rosalind_8614 Rosalind_1595\n",
      "Rosalind_4448 Rosalind_0539\n",
      "Rosalind_8520 Rosalind_0716\n",
      "Rosalind_4044 Rosalind_7161\n",
      "Rosalind_4044 Rosalind_8999\n",
      "Rosalind_7430 Rosalind_6432\n",
      "Rosalind_1009 Rosalind_2535\n",
      "Rosalind_1009 Rosalind_1808\n",
      "Rosalind_1009 Rosalind_4575\n",
      "Rosalind_7285 Rosalind_6992\n",
      "Rosalind_7285 Rosalind_4261\n",
      "Rosalind_7285 Rosalind_5499\n",
      "Rosalind_8041 Rosalind_6798\n",
      "Rosalind_1885 Rosalind_2463\n",
      "Rosalind_6992 Rosalind_1213\n",
      "Rosalind_4261 Rosalind_6438\n",
      "Rosalind_4261 Rosalind_0707\n",
      "Rosalind_3229 Rosalind_6992\n",
      "Rosalind_3229 Rosalind_4261\n",
      "Rosalind_3229 Rosalind_5499\n",
      "Rosalind_3349 Rosalind_7682\n",
      "Rosalind_5934 Rosalind_0716\n",
      "Rosalind_2898 Rosalind_4448\n",
      "Rosalind_3479 Rosalind_7955\n",
      "Rosalind_3479 Rosalind_9492\n",
      "Rosalind_3479 Rosalind_4859\n",
      "Rosalind_6715 Rosalind_1009\n",
      "Rosalind_0602 Rosalind_6798\n",
      "Rosalind_0117 Rosalind_2890\n",
      "Rosalind_0247 Rosalind_3479\n",
      "Rosalind_0247 Rosalind_1565\n",
      "Rosalind_0247 Rosalind_7216\n",
      "Rosalind_6766 Rosalind_5934\n",
      "Rosalind_6766 Rosalind_2447\n",
      "Rosalind_9413 Rosalind_2465\n",
      "Rosalind_4914 Rosalind_0247\n",
      "Rosalind_4914 Rosalind_3419\n",
      "Rosalind_4914 Rosalind_9005\n",
      "Rosalind_8465 Rosalind_7430\n",
      "Rosalind_8465 Rosalind_3517\n",
      "Rosalind_8465 Rosalind_1735\n",
      "Rosalind_8465 Rosalind_4660\n",
      "Rosalind_2320 Rosalind_8614\n",
      "Rosalind_2320 Rosalind_6715\n",
      "Rosalind_2320 Rosalind_6766\n",
      "Rosalind_0861 Rosalind_2898\n",
      "Rosalind_0861 Rosalind_1401\n",
      "Rosalind_0861 Rosalind_2188\n",
      "Rosalind_2535 Rosalind_0444\n",
      "Rosalind_3406 Rosalind_9785\n",
      "Rosalind_3406 Rosalind_4234\n",
      "Rosalind_6438 Rosalind_4559\n",
      "Rosalind_6438 Rosalind_7441\n",
      "Rosalind_6438 Rosalind_3214\n",
      "Rosalind_0539 Rosalind_2963\n",
      "Rosalind_0539 Rosalind_4169\n",
      "Rosalind_5499 Rosalind_6798\n",
      "Rosalind_6702 Rosalind_2463\n",
      "Rosalind_1595 Rosalind_5934\n",
      "Rosalind_1595 Rosalind_2447\n",
      "Rosalind_9492 Rosalind_9516\n",
      "Rosalind_9492 Rosalind_6424\n",
      "Rosalind_9492 Rosalind_0344\n",
      "Rosalind_2483 Rosalind_8728\n",
      "Rosalind_6238 Rosalind_7682\n",
      "Rosalind_3419 Rosalind_1213\n",
      "Rosalind_5506 Rosalind_9419\n",
      "Rosalind_5506 Rosalind_0585\n",
      "Rosalind_4559 Rosalind_3479\n",
      "Rosalind_4559 Rosalind_1565\n",
      "Rosalind_4559 Rosalind_7216\n",
      "Rosalind_7441 Rosalind_0303\n",
      "Rosalind_4169 Rosalind_1213\n",
      "Rosalind_3517 Rosalind_0303\n",
      "Rosalind_9419 Rosalind_2320\n",
      "Rosalind_3364 Rosalind_6432\n",
      "Rosalind_3563 Rosalind_8465\n",
      "Rosalind_9046 Rosalind_5506\n",
      "Rosalind_2447 Rosalind_1885\n",
      "Rosalind_2447 Rosalind_4062\n",
      "Rosalind_7161 Rosalind_1595\n",
      "Rosalind_3280 Rosalind_1213\n",
      "Rosalind_9516 Rosalind_1595\n",
      "Rosalind_4575 Rosalind_2463\n",
      "Rosalind_1213 Rosalind_9201\n",
      "Rosalind_1401 Rosalind_8520\n",
      "Rosalind_1401 Rosalind_2483\n",
      "Rosalind_8126 Rosalind_6992\n",
      "Rosalind_8126 Rosalind_4261\n",
      "Rosalind_8126 Rosalind_5499\n",
      "Rosalind_0190 Rosalind_6238\n",
      "Rosalind_4062 Rosalind_2320\n",
      "Rosalind_6424 Rosalind_0303\n",
      "Rosalind_1565 Rosalind_7430\n",
      "Rosalind_1565 Rosalind_3517\n",
      "Rosalind_1565 Rosalind_1735\n",
      "Rosalind_1565 Rosalind_4660\n",
      "Rosalind_9792 Rosalind_0117\n",
      "Rosalind_9005 Rosalind_1885\n",
      "Rosalind_9005 Rosalind_4062\n",
      "Rosalind_1378 Rosalind_6438\n",
      "Rosalind_1378 Rosalind_0707\n",
      "Rosalind_4642 Rosalind_2535\n",
      "Rosalind_4642 Rosalind_1808\n",
      "Rosalind_4642 Rosalind_4575\n",
      "Rosalind_9785 Rosalind_2535\n",
      "Rosalind_9785 Rosalind_1808\n",
      "Rosalind_9785 Rosalind_4575\n",
      "Rosalind_7153 Rosalind_1213\n",
      "Rosalind_0344 Rosalind_1885\n",
      "Rosalind_0344 Rosalind_4062\n",
      "Rosalind_8999 Rosalind_7232\n",
      "Rosalind_8675 Rosalind_9419\n",
      "Rosalind_8675 Rosalind_0585\n",
      "Rosalind_0707 Rosalind_4448\n",
      "Rosalind_2890 Rosalind_6432\n",
      "Rosalind_7232 Rosalind_5934\n",
      "Rosalind_7232 Rosalind_2447\n",
      "Rosalind_6432 Rosalind_0716\n",
      "Rosalind_4859 Rosalind_0247\n",
      "Rosalind_4859 Rosalind_3419\n",
      "Rosalind_4859 Rosalind_9005\n",
      "Rosalind_2188 Rosalind_8465\n",
      "Rosalind_2188 Rosalind_3563\n",
      "Rosalind_1735 Rosalind_7232\n",
      "Rosalind_4334 Rosalind_7161\n",
      "Rosalind_4334 Rosalind_8999\n",
      "Rosalind_2465 Rosalind_7430\n",
      "Rosalind_2465 Rosalind_3517\n",
      "Rosalind_2465 Rosalind_1735\n",
      "Rosalind_2465 Rosalind_4660\n",
      "Rosalind_4234 Rosalind_1885\n",
      "Rosalind_4234 Rosalind_4062\n",
      "Rosalind_7216 Rosalind_0117\n",
      "Rosalind_7216 Rosalind_9792\n",
      "Rosalind_6596 Rosalind_6992\n",
      "Rosalind_6596 Rosalind_4261\n",
      "Rosalind_6596 Rosalind_5499\n",
      "Rosalind_8728 Rosalind_2465\n",
      "Rosalind_0716 Rosalind_8041\n",
      "Rosalind_0716 Rosalind_3229\n",
      "Rosalind_9307 Rosalind_9785\n",
      "Rosalind_9307 Rosalind_4234\n",
      "Rosalind_4923 Rosalind_0716\n"
     ]
    }
   ],
   "source": [
    "\n",
    "import networkx as nx\n",
    "\n",
    "k = 3\n",
    "\n",
    "# Get contents of FASTA file\n",
    "dict = read_FASTA_file(open('question_inputs/graph_overlap_input.txt', 'r'))\n",
    "\n",
    "digraph = nx.DiGraph()\n",
    "digraph.add_nodes_from(dict.keys())\n",
    "\n",
    "pairs = [(s,t) for s in dict.items() for t in dict.items() if s!=t]\n",
    "\n",
    "#for each pair, if prefix matches suffix, create edge. \n",
    "# direct edge from suffix to prefix string. \n",
    "for s,t in pairs:\n",
    "    if s[1][:k] == t[1][-k:]:\n",
    "        digraph.add_edge(t[0], s[0])\n",
    "\n",
    "for e in digraph.edges():\n",
    "    print(e[0] +\" \"+e[1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculating Expected Offspring\n",
    "\n",
    "Molecular biology research uses averages to predict the expected number of antibiotic-resistant pathogenic bacteria in a future outbreak, estimate the predicted number of locations in the genome taht will match a given motif, and to study the distribution of alleles throughout an evolving population.\n",
    "\n",
    "### Problem\n",
    "\n",
    "Given 6 integers that represent the number of couples in a population each possessing each genotype pairing for a given factor, return the expected number of offspring displaying the dominant phenotype in the next generation, assuming that every couple has two offspring.\n",
    "\n",
    "In order, the six values correspond to the genotype:  \n",
    "AA-AA  \n",
    "AA-Aa  \n",
    "AA-aa  \n",
    "Aa-Aa  \n",
    "Aa-aa  \n",
    "aa-aa  \n",
    "\n",
    "\n",
    "### Solution\n",
    "\n",
    "For each population, the probabilty that one offsping has the dominate trait is given below. To compute the expected value we apply the formula, and multiple by two as there are two offsprings (assuming independence.)\n",
    "\n",
    "AA-AA -> 1  \n",
    "AA-Aa -> 1  \n",
    "AA-aa -> 1  \n",
    "Aa-Aa -> 0.75  \n",
    "Aa-aa -> 0.5  \n",
    "aa-aa -> 0  \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "158802.5\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "pop = np.array([19004, 16403, 19730, 19939, 18620, 16352])\n",
    "\n",
    "prob = np.array([1, 1, 1, 0.75, 0.5, 0])\n",
    "\n",
    "print(np.sum(pop*prob*2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Finding a Shared Motfi\n",
    "\n",
    "Recall a **motif** is a substring of nucleotides or amino acids, yet this motif may not be known in advance. \n",
    "\n",
    "A common motif to find is the longest common substring of a collection of genetic strings. A longer motif will likely indicate a greater shared function between the genetic strings.\n",
    "\n",
    "### Problem\n",
    "Given a collection of DNA strings in FASTA format, determine the longest common substring (LCS) of the collection. \n",
    "\n",
    "TODO see generalized suffix trees. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get contents of FASTA file\n",
    "dict = read_FASTA_file(open('question_inputs/lcs.txt', 'r'))\n",
    "\n",
    "A = list(dict.values())[0]\n",
    "B = list(dict.values())[1:]\n",
    "\n",
    "# TODO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
