{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Rosalind Problems - Bioinformatics Stronghold"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Counting DNA Nucleotides (nt)\n",
    "\n",
    "The nucleus of any living cell consists of macromolecules called chromatin. One class of marcomolecules in the chormatin are nucleic acids. Nucleic acids are polygmers - repeating chains of small, similarly strucutred molecules called monomers (stands.)\n",
    "\n",
    "The **nucleotide (nt)** is the unit of stand length and is composed of three parts: a sugar molecule, a negatively charged phosphate ion, and a compound called a nucleobase (base.) THe sugar of one nucleotide bonds to the the phosphate of the next nucleotide in the chain, forming a sugar-phosphate backbone.\n",
    "\n",
    "Nucleotides of a specific type of nucleic acid always contains the same sugar and phosphate molecule; thus they only differ by their choice of base. Therefore a strand of nucleic acid can be differentiated from another based only on the order of its bases - this ordering defines a nucleic acid's **primary structure**. \n",
    "\n",
    "For a strand of **deoxyribose nucleic acid (DNA)**, the four necleobases are molecules called **adenine (A)**, **cytosine (C)**, **guanine (G)**, and **thymine (T)**."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem: \n",
    "\n",
    "Given a string over the alphabet {A,C,G,T}, count the number of times each symbol occurs in the string."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_values([247, 244, 242, 242])\n"
     ]
    }
   ],
   "source": [
    "input_string = \"ATTCGTAGGCATCGCTCCTGCCCTGTGGTTACGTTCGCACCGGGTGTCAGGTCTGAAATTATGAGAATTTTACACTCCTAATGTTGGGCGACTGAACTGATTTTGAAGTCAGGCGGTAGTATACTGAGTCAGGGTCGGAGAGTGAGAGATGTATGAGCGTGGTTTGCTCCGCTAGCCCAGGCCGGCGAGTGCTGATGCGTAGCTGTATCTACAATTCGTCAGCTTTTGGCGCAGTAATGAATATCAGAGTGCCTAAGCAGCTCTTGCCCTACGGCTAATCGAATCTTCAGATTCAAAGGGGTACACTAGCTGTTGCCAAACCGTCGGAGACGCCACCTCACGAACTCACATCATATCAGTCGAAGATGGTACGTTAGAGCCCATGTGCCTATGGAGGCATGGTATATAGCCTGGCTGCGGTGTAATGAAGTCACCTGGCCAACCTGGGGCACTTGTCTCGTGACATAGTCTTATAGGGGTGCCGCAAGAAGGTCCCACCAAAAATCACGGATACGCGTTCGCCCGAAGGCCCTTGGAAACCTTCCAGCTCAAACTTGAGAGATACGTAGTTTTCCCACCCGTCACAAGAGGTGAGATTACAACAAAGTATGTCTGCTCGTCCCTGTTAGCTATAGCAGCTCGTATAACAACATTTCCCGACTGCGCGTCGCTGATCGAACAATAGGCTAGAGTTACAGAGCCTATCACAGGCCGAAGAGGACAAGCTTCGCTGGTACCTATAAAGGGAAAACCCCAGATTAATCCCGCCTTATCGTCCTAAGTAGTCGCGGTCAATATGGTACTATCCGACACCCTTTGATTGTGAGTCTTCACTTACCTATTCCGGATTTCCTCTCCCAGTAGGTTATAGTTCATGACGTCTTTTAGCACGGGATCACACGTGGCATGGATTAAATAGGATACACCTGGTGCCATGCAAAGTCCACGCTTAAGCCGAAGAACCCCGTAAAGCAC\"\n",
    "\n",
    "# Create dictonary for each character, scan through string, counting occurences.\n",
    "dict = {'A':0, 'C':0, 'G':0, 'T':0}\n",
    "\n",
    "for char in input_string:\n",
    "    count = dict[char]\n",
    "    dict[char] = count + 1\n",
    "    \n",
    "print(dict.values())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## The Second Nucleic Acid (RNA)\n",
    "\n",
    "Along side DNA in the chromatin a molecule with a different sugar can ribose known as **ribose nucleic acid (RNA)**. RNA differs from DNA as in it contains a base called **uracil** instread of thymine. \n",
    "\n",
    "The primary structure of DNA and RNA is similar because DNA serves as the blueprint for the creation of special kinds of RNA molecule called **messenger RNA (mRNA)**. mRNA is created during RNA transcription, during which a strand of DNA is used as a template for constrcuting a strand of RNA by compy nucleotides one at a time, where uracil is used in place of thymine."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem: \n",
    "\n",
    "Given a DNA string t, convert t to RNA by replacing all thymine (T) with uracil (U)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AAAACGUUGCCCUCCUCCCAGGUUCUGGUAAACGUGCACAGCGUCCCCUGCCUUUCAUAUCGUACUUCUAGGCCCUCGUAAGGAGUCCGGUAUGUAAUAUCGAUCACGAUCGCCUAUGAGGUGCGGGAAUUGCCAAAUCACUGGCCUAGUUGACUCAGUCACCCUCUUUUGCAAGGCCUUUAGCGAGGAAACCUGCACCGCCGAGGUACGCAAUUGCUACAUGCUACCCAAGAAGGGCAUUCAGCACGCGAGAGCGAUACUGUACUCGAGCACCACCAUCACGAUAGAAAGGAUGGUUGUAUUCACACGCGGUUAGGUGGAACCUUGAUAAAUGUGCGCGGAUGCGAAUGUGCCAAGUUGCCUGCUCUUUCUUAAGGUUCCGGUCUGACGGGAACACAAGGCGAGUAAGAGGGUCUUAGCAAGGGGUGCGCCUUAGCGACUACUAUUAGCUUGCACACAAACAAGCGAGUAGUAUGGCCUCGCUCGUGUUCAGCUCGGUUUGCGCGUGACCAUUGCAUUUUUUAAAUUGAAAAACCAGUGGUGCUGCUCACUAGCAUUAUCCGUUUGCUAGCCCUUCCAAUGUAACAAUGCAGUUGGGCUGAAGAGAGACCAGUACGCCCAGCUAUUCUAGAAUCCACCUCUGGAAGAACAACUGAGUAUUCUCGAAGUUUACCAGCAUCCCUCAGUAAAAUUUCAACUAUGAUGUUGUCAUCCAAUUGCUCCCGAGAAAAUUCGACGUGUAGCCCGAUGAAAGGUAAGUCGAGUGGGAUACCUCUAAUGAUAGUGCAGAGAAUGGCCCGUAGCCUCGGCACUAGAGCCCGUUGCCACUAAGAGCCAAGAAAAGCCCUCGUGAGCAGAGCCUCCGUGCAGUCUCUCUGGCCGGCAAUGCAAGGACAGGCUGGUUGC\n"
     ]
    }
   ],
   "source": [
    "t = \"AAAACGTTGCCCTCCTCCCAGGTTCTGGTAAACGTGCACAGCGTCCCCTGCCTTTCATATCGTACTTCTAGGCCCTCGTAAGGAGTCCGGTATGTAATATCGATCACGATCGCCTATGAGGTGCGGGAATTGCCAAATCACTGGCCTAGTTGACTCAGTCACCCTCTTTTGCAAGGCCTTTAGCGAGGAAACCTGCACCGCCGAGGTACGCAATTGCTACATGCTACCCAAGAAGGGCATTCAGCACGCGAGAGCGATACTGTACTCGAGCACCACCATCACGATAGAAAGGATGGTTGTATTCACACGCGGTTAGGTGGAACCTTGATAAATGTGCGCGGATGCGAATGTGCCAAGTTGCCTGCTCTTTCTTAAGGTTCCGGTCTGACGGGAACACAAGGCGAGTAAGAGGGTCTTAGCAAGGGGTGCGCCTTAGCGACTACTATTAGCTTGCACACAAACAAGCGAGTAGTATGGCCTCGCTCGTGTTCAGCTCGGTTTGCGCGTGACCATTGCATTTTTTAAATTGAAAAACCAGTGGTGCTGCTCACTAGCATTATCCGTTTGCTAGCCCTTCCAATGTAACAATGCAGTTGGGCTGAAGAGAGACCAGTACGCCCAGCTATTCTAGAATCCACCTCTGGAAGAACAACTGAGTATTCTCGAAGTTTACCAGCATCCCTCAGTAAAATTTCAACTATGATGTTGTCATCCAATTGCTCCCGAGAAAATTCGACGTGTAGCCCGATGAAAGGTAAGTCGAGTGGGATACCTCTAATGATAGTGCAGAGAATGGCCCGTAGCCTCGGCACTAGAGCCCGTTGCCACTAAGAGCCAAGAAAAGCCCTCGTGAGCAGAGCCTCCGTGCAGTCTCTCTGGCCGGCAATGCAAGGACAGGCTGGTTGC\"\n",
    "print(t.replace('T', 'U'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## The Secondary and Tertiary Structures of DNA\n",
    "\n",
    "The primary structure of a nucleic acid is determined by the ordering of its bases, yet it does not describe the large, 3D shape of the molecule. In 1953, the following structure for DNA was proposed:\n",
    "\n",
    "1) DNA is composed of two strings, running in opposite directions. \n",
    "\n",
    "2) Each base bonds to a base in the opposite strand: A-T, C-G (always)\n",
    "\n",
    "3) The two strands are twisted together into a long spiral double helix. \n",
    "\n",
    "\n",
    "1) and 2) compose the **secondary structure** of DNA. 3) Describes the **tertiary structure**.\n",
    "\n",
    "Note that the **complement** of a base is the base in which it bonds to. The bonding of two complementary bases is called a **base pair (bp)**. Thus the length of a DNA molecule is commonly given in bp rather than nt; we can determine the other strand by taking the complement of the first (running in opposite direction.)\n",
    "\n",
    "Example:\n",
    "\n",
    "'AAAACCCGGT' <-> 'ACCGGGTTTT'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TATAAACCGTGGGAAACGTCGATCCGCAACGACGAAAACTTTAGAAGGATCCTTATCTAATAGTGGTCCACGTTCCCCCTGTGCATTAACATAAAAAGAACCCTGACGTGATCCTCTAAACCTGGTACTACATTTTCACAGCTGGTTACACCGAGTGCTTAGCTTACGTCAGTACACGGATGAAAGGCACTGTCCTATACCTATCTTGAGTAGAGGGCAAGAATGGAAATTATACTCCGTTAGCTAAGGTATCTCTATAAGACATTAGGTTGTACCAAAAATGCAGCATTCAGGCGGGGTGCAGGGCTTGCCGGGGAACTGCTCGGATGGCTTGTAAGACAACACCCGCTCGAAGGTTCATGCAAACCTTACAATAACCGGAGACTCAGCTCGAAGAATTGCCCAATTTCGCCCCCTACCGCTGATTTGCGGCGATATGGGCCCTTCATATAATACAGAGTCTGCGCTACGAGACCTCAGTAGCCATCCCTAAATCGGTGGAGCGGCTTCTGACATTGGATCTCGAGCACTCTAGTCGACCGTTACTTTCAGAGCAATGCCTAAAAATGTCAAGCGCCGATGCGAAAGGATCCCATACACTGTAGTGAGTAGGTATGGACCTGGACGATACTAACGCTAGGCAGTCCCCTCAAGGGGCGCACCAGATCGCGTTGTCGTGAAGCGGCTAAGGGTTCCCCAGCCTTTACGAGACACAAAATGGCTGCCTAGAGTCATGTGGCCCGTTCTCTATCCCCAACTATCGTGTAAGATCGAGGAAAGTTGGGGTACGGCGGCGCCTCCTACGTGGGCGCCTCGTAACGATAGAGCCCCTGTCTGGAGTT\n"
     ]
    }
   ],
   "source": [
    "s = \"AACTCCAGACAGGGGCTCTATCGTTACGAGGCGCCCACGTAGGAGGCGCCGCCGTACCCCAACTTTCCTCGATCTTACACGATAGTTGGGGATAGAGAACGGGCCACATGACTCTAGGCAGCCATTTTGTGTCTCGTAAAGGCTGGGGAACCCTTAGCCGCTTCACGACAACGCGATCTGGTGCGCCCCTTGAGGGGACTGCCTAGCGTTAGTATCGTCCAGGTCCATACCTACTCACTACAGTGTATGGGATCCTTTCGCATCGGCGCTTGACATTTTTAGGCATTGCTCTGAAAGTAACGGTCGACTAGAGTGCTCGAGATCCAATGTCAGAAGCCGCTCCACCGATTTAGGGATGGCTACTGAGGTCTCGTAGCGCAGACTCTGTATTATATGAAGGGCCCATATCGCCGCAAATCAGCGGTAGGGGGCGAAATTGGGCAATTCTTCGAGCTGAGTCTCCGGTTATTGTAAGGTTTGCATGAACCTTCGAGCGGGTGTTGTCTTACAAGCCATCCGAGCAGTTCCCCGGCAAGCCCTGCACCCCGCCTGAATGCTGCATTTTTGGTACAACCTAATGTCTTATAGAGATACCTTAGCTAACGGAGTATAATTTCCATTCTTGCCCTCTACTCAAGATAGGTATAGGACAGTGCCTTTCATCCGTGTACTGACGTAAGCTAAGCACTCGGTGTAACCAGCTGTGAAAATGTAGTACCAGGTTTAGAGGATCACGTCAGGGTTCTTTTTATGTTAATGCACAGGGGGAACGTGGACCACTATTAGATAAGGATCCTTCTAAAGTTTTCGTCGTTGCGGATCGACGTTTCCCACGGTTTATA\"\n",
    "s_c = \"\" # complement of s. \n",
    "\n",
    "for char in s:\n",
    "    if char == 'A':\n",
    "        s_c += 'T'\n",
    "    elif char == 'T':\n",
    "        s_c += 'A'\n",
    "    elif char == 'C':\n",
    "        s_c += 'G'\n",
    "    elif char == 'G':\n",
    "        s_c += 'C'\n",
    "        \n",
    "print(s_c[::-1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Rabbits and Recurrences\n",
    "\n",
    "We begin with a pair of rabbits - one male and one female. Each month such a pair mates, and one month later k pairs are produced (one male, one female.)\n",
    "After a month of being born, rabbits reach reproductive age. Rabbits never stop reporducing or dying. \n",
    "\n",
    "Given n months and value k, what is the the total number of rabbit pairs after n months?\n",
    "\n",
    "e.g. n=5, k=3 -> 19 pairs. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Recurrence relationship:\n",
    "\n",
    "Month m=0, there is one pair of mating age, 0 new born. Total: 1\n",
    "Month m=1, there is one pair of mating age, k pairs of new borns. Total: k+1\n",
    "Month m=2, there are k+1 pairs of mating age, k pairs of new borns: Total: 2k+1\n",
    "\n",
    "In general, for the ith month suppose there are x,y pairs of mating age, new born respectively. Then for month i+1, we have\n",
    "x+y,k(x) respectvely. \n",
    "\n",
    "Let $F_i$ denote the number of pairs of rabbits on the $i^{th}$ month. Then for $i=0, i=1$, we have $F_0=1$, $F_1=k+1$, and $F_i = F_{i-1} + kF_{i-2}$ for $i \\geq 2$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def recurrence(m,k):\n",
    "    if m == 0:\n",
    "        return 0\n",
    "    elif m == 1:\n",
    "        return 1\n",
    "    else:\n",
    "        return recurrence(m-1, k) + k * recurrence(m-2, k) \n",
    "# See a bottom up solution."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3048504677680\n"
     ]
    }
   ],
   "source": [
    "m = 36\n",
    "k = 3\n",
    "print(recurrence(m, k))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GC-Content\n",
    "\n",
    "99.9% of the 3.2 billion base pairs in a human genome are common to almost all other humans (barring those with mahor genetic defects.) For any species, the average case genome is very similar.\n",
    "\n",
    "For DNA, this means that the frequencey of bases will be similar for memebers of the same species. \n",
    "\n",
    "For a molecule of DNA that is from an unknown species, we can analysis the frequencies of bases in the molecule and compare them to a database of DNA to attempt to determine the species. \n",
    "\n",
    "Because of the base pairing relations of the two DNA strands, cytosine and guanine will always appear in equal amounts in a double-stranded DNA molecule - this we analysis its **GC-Content**. The GC-content of a DNA string is the percentage of symbols in the string that are either 'C' or 'G'.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### FASTA format\n",
    "\n",
    "The FASTA format is a means to label strings (DNA) In this format, we introduce the label of the string using a '>' followed by the label. Then the next line is the string itself. For example,\n",
    "\n",
    "\\>Rosalind_6404\n",
    "CCTGCGGAAGATCGGCACTAGAATAGCCAGAACCGTTTCTCTGAGGCTTCCGGCCTTCCCTCCCACTAATAATTCTGAGG\n",
    "\n",
    "\\>Rosalind_5959\n",
    "CCATCGGTAGCGCATCCTTAGTCCAATTAAGTCCCTATCCAGGCGCTCCGCCGAAGGTCTATATCCATTTGTCAGCAGACACGC\n",
    "\n",
    "\\>Rosalind_0808CCACCCTCGTGGTATGGCTAGGCATTCAGGAACCGGAGAACGCTTCAGACCAGCCCGGAC\n",
    "TGGGAACCTGCGGGCAGTAGGTGGAAT\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem:\n",
    "\n",
    "Give a list of FASTA formatted DNA, compute the string with largest GC-content."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Rosalind_6489', 0.5282583621683967]\n"
     ]
    }
   ],
   "source": [
    "# Get input from file, store content in list of pairs: label, string\n",
    "file = open('question_inputs/computeGC.txt', 'r')\n",
    "\n",
    "# preprocessing\n",
    "dna_list = file.read().replace('\\n', '').split('>')\n",
    "dna_list.pop(0)\n",
    "\n",
    "# For each pair, compute GC-content of string, store label and freq.\n",
    "largest_pair = [None, -1.0]\n",
    "\n",
    "i = 0\n",
    "while (i < len(dna_list)):\n",
    "    label = dna_list[i]\n",
    "    seq = dna_list[i+1]\n",
    "    # Compute GC-content\n",
    "    count = 0\n",
    "    for char in seq:\n",
    "        if char == 'G' or char == 'C':\n",
    "            count += 1\n",
    "    if largest_pair[1] < count/len(seq):\n",
    "        largest_pair = [label, count/len(seq)]\n",
    "    \n",
    "    i += 2\n",
    "    \n",
    "print(largest_pair)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Point Mutations\n",
    "\n",
    "A **mutation** is an error that occurs during the creation or copying of a nucleci acid (in particular DNA.) Mutations come in forms, the simplest and most common is a **point mutation** which replaces one base with another at a single nucleotide. For DNA, this also changes the complementary base.\n",
    "\n",
    "For example, the pair G-C mutations to A-T.\n",
    "\n",
    "The **Hamming distance** between two strings of equal length is the number of points in which they differ. \n",
    "\n",
    "DNA from two species with a common ancestor are expected to have the similar strings (i.e. small Hamming difference.)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem: \n",
    "\n",
    "Given two strings of the same length, compute the Hamming distance between the two."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "446\n"
     ]
    }
   ],
   "source": [
    "s = \"GAGCCTACTAACGGGAT\"\n",
    "t = \"CATCGTAATGACGGCCT\"\n",
    "\n",
    "s = \"CACATAACAGCCCTGCAACCACCTAGGCCTGGAGCTACTGCTGCCAGCACTCCTGCAGTGAACTTGAAATACTTCGATGCATAATGAACTTTGTCTATTTCGGGCGCACGCAACACCGACAGCTACGATCTTCATCGAAGTTTTAGCGCCAAGATAGTTACTGCTTGTAGCCAGGTGGCGACGCTTAATAGCACGAGTGACCCTGAGGACAAATGTTGCCTCAGGTCATCAAGACGCGGCGGCTATTACTCAGTCAAACCGGCATGGGTCCCACGACCTGATACATCTCCGACTGGATGATATTATCAGATTAGCCGCAGCGTATCAAGGAACTATTCAGTAGTACTGCCTCGGGGGGTATGAACGCACGAGGCATTTAATAGTTAATGCTAACCACACTTCCAAATCATATGAAGTGGGACAGTTAGCAATGATGGCACTGGAAGAGTTCACAGTTCCTCTAAAAAGAACTAGGGTCCGGACATGGACGCTTTCGGTCGGATTTACTGCGAGTTCAAATTGGATGAAGTGGCCCACCGAATAGGTTCGAGCAACTTACTCGCAGCTATAAGGGAGTGTTATACATGTATATTTCCCCTTTTAAGCCCGTACGTGGCTCCAGATATAGACTCCTTGCAGGAAGGTGTGCACCCCTCGCTGTGCACGTGGCCGTAGATCTCGGCCCGTCCATGAGGCGATCAGTGCCAGGTGCTTTAGCCTTGTGCTGTCGGCACGTCTAGCTCCGATATGGGTCCCCTTGCCTTTTACTCACCGAAACGCGGTCGGGGGGGGTTGGTTCGGTGCCAAGGGTAAAGGGTAGCTCGAGATGTAGCGCCGTAAAACAAACGAATGGGTAATGGTATCTCCTGCAGTCTAGGTCCTATGAGCCTCTAGTGAGTATCTCGTCTCATCCACGTGCGTCAGCATAACTT\"\n",
    "t = \"CTATTAACTTATTGGCAACCAAGTACGGCTAGTGCTACAGAAACCACTAATCCTAAACTGCACCTTAAATATTTGCATGTACAACAAGTGTAGTACACATGATGCACAAGCGAGCGCGGGAGTTCGGAGCACCGATTATTATTCAGCGCGAGGAAAACCTCTTCGTTCGGTGAAGTGGCGGGGTTTTATGCCAGGTAACGGAATGACGGCAAATATACCGCCCGGTGTCCTACTTTATGCTACAACTACAATTTCGTACCGGCATGGGTCTCTCCGGCTAAGCCATTGCCCATTGGAACGTATAGCAGCGGAAGCCATATCGTATCAAAATTCTAGCCTGGGAGGCCGCCTCAGGCACAATGGACACCCGAGGTATCTAGTAGTTGGTCCTATGCACACTTCAGTACGCTTTGACACACGACAGATAGTCGAGATGGCCGAGATACGCCGCAGCTTTGCTTTAATGAGAGCTAGGTCCCGTGTTCGGATACGTTAGGCCATATTAGGAACACGTCGCACATGGATCCCCTGGAGAACCATATCGGTGTTAGCCTCTTGGCGGCCGCTGCAGAGCACTGTTGTACGTGAACATGTTCTATACCGGGCTCGTACATTACTGCGAGTCTATCTTTGGTGTAGTAGATTAGGCTTACCTAGCATTACACTCTCACCAAGATAGCTGCTACCCCAATAGTGGAAACGGACAACGGGTTTCAGTCGGGTGCCGACCGCACGGATGACATACATACGTCACCGCTCGAGGTTGACGAGTCGTAGCAAGCTCGGTAGGCCTTTTTGACGACGGAAGTGATGAGTACAGCTGTTCAATTAGAAGCTGCCCACTAAGGTGGATGTAAATCTAGCCGAGTCAGTGCAAGCCCGGCGAGGCACCGGCGAGTCTCTCGTCTAAAACAGGTGCGGCGGAGCAGCAT\"\n",
    "\n",
    "\n",
    "ham_dst = 0\n",
    "for i in range(len(s)):\n",
    "    if s[i] != t[i]:\n",
    "        ham_dst += 1\n",
    "        \n",
    "print(ham_dst)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mendelian Inheritance\n",
    "\n",
    "In 1865, Gregor Mendel described the contemporary hereditary **blending inheritance** model, which states that the inheritance must be a blend of its partent's traits. This model however is not observed in people.\n",
    "\n",
    "Mendel viewed the traits of pea plants as discrete blocks called **factors**, each of which has a distinct form called **alleles**.\n",
    "\n",
    "Mendel's **first law (law of segregation)** states that for a given factor, every organism possesses a pair of alleles for a given factor. If an individual's two allels for a given factor are the same, then it is **homozygous** for the factor; if they differ, then the individual is **heterozygous**. Thus for an offspring with two parents will receive one allele from each parent; which one is randomly decided. \n",
    "\n",
    "Further, any factor corresponds to only two possible alleles - the **dominant** and **recessive** alleles. By possessing only one copy of the dominate allele, the organism will **display the trait** of the dominant allele. That is, the only way that an organism can display a trait encoded by a recessive allele is if the individual is homozygous recessive for that factor.\n",
    "\n",
    "We encode the dominant allele of a factor using captial letters, and the recessive allele by lower case letters. As an organism may possess the recessive allele but not display it, we define an organism's **genotype** as its precise genetic makeup, and its **phenotype** as its physical manifestation of its underlying traits.\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem\n",
    "\n",
    "Given populations K,M,N of homozygous dominant, hetrozygois, and homozygous recessive factors respectively, compute the probability of any two mating organisms will produce an individual possessing a dominant allele. (An organism cannot mate with itself.) Assume at least two memebers of each population.\n",
    "\n",
    "### Solution\n",
    "\n",
    "From k,m, and n numbers in each population, we consider all events (i.e. ways to mate) along with the probability that the resulting individual will possess the dominant allele:\n",
    "\n",
    "Parents:ways that offspring can have the dominante allele.\n",
    "\n",
    "KK:4/4  \n",
    "KM:4/4  \n",
    "KN:4/4  \n",
    "\n",
    "MK:4/4  \n",
    "MM:3/4  \n",
    "MN:2/4  \n",
    "\n",
    "NK:4/4  \n",
    "NM:2/4  \n",
    "NN:0/4  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.8364182692307692\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "k=29\n",
    "m=19\n",
    "n=17\n",
    "sum_terms = k+m+n\n",
    "total_outcomes = sum_terms*(sum_terms-1)\n",
    "\n",
    "dominante_allele_outcomes = [\n",
    "    k*(k-1)*1,\n",
    "    k*m*1,\n",
    "    k*n*1,\n",
    "    \n",
    "    m*k*1,\n",
    "    m*(m-1)*0.75,\n",
    "    m*n*0.5,\n",
    "    \n",
    "    n*k*1,\n",
    "    n*m*0.5,\n",
    "    n*(n-1)*0\n",
    "    \n",
    "]\n",
    "\n",
    "print(np.sum(dominante_allele_outcomes) / total_outcomes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Translating RNA into Protein\n",
    "\n",
    "**Proteins** are chains of small molecules called **amino acids** - 20 amnio acids commonly appear in all species. Similar to nucleic acids, the **primary structure of a protein** is the order of the amnio acids.\n",
    "\n",
    "Proteins power all practical functions carried out by a cell; thus the key to understanding life lies in understanding the relationship between a chain of amino acids and the function of the protein that this chain will eventually construct. \n",
    "\n",
    "The translation of an RNA molecule (mRNA) into amino acids are how amino acids are created. Somehow, the 4 RNA bases must be translated into a language of **20 amino acids**; for every possible amino acid, we translate 3 nucleobase strings (called **codons**) into amino acids - this implies there are $4^3=64$ possible codons.  \n",
    "Two special types of codons are the **start codon (AUG)** and three **stop codons (UAA, UAG, UGA)**, these do not code for amnio acid, only indicate the translation to end.\n",
    "\n",
    "#TODO update"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem \n",
    "Given an input RNA string s corresponding to a strand of mRNA, output the protein string encoded by s. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MCHIVTRYSAHRPVHAELQFASSPTRSSDGGYKSLKAMGTTVLLRSLRSISCRTYYIIPCRSPRLEHFRLSCKISLFHARQTMLKVTVMTYPAKIAACARRAKLPFILCGGMVPSCSRVMKYGCIRYRETGKQLKCVKTALRPTNENSSIARAKCNVNVKGWLRYQLCADQGMAQLRACPMGRRGRSDMRLIGPRVLLPSGGCSKSGWYNLSLTDRPLPVSCLPMGSFKSASSNITGASPCMTIMPIIKTQNFAHCMPKLTPRITQQLKRLLLPPVFAPARPRGHSHVYQTGIAFCTPWTPEPYQSRCCHGLLAVLLIRRTFTSLQPSRFRNNIAEVSLSESIWAYCYHIEPYSLELYPEMGHASTSVSRLRPRQVAVRVRANRRISLRVQTQCGKLPTGLWVPIGKSNGDTLQPWVEVDKKFLGERTDTVETYLEICIVRPEGFSRSVELQRCDFRFWTQVSTQVTKPPVKAAVCIYNVSRRSYLSAPSNPSLPLFSQRCCNWGFSTHSEMPGDDLSPLLSVKHLHMGLRVLEAGYQDSKRHYYTTCRESSLYLRIALAVVARRKVRHHSQDDTKSLPWNNSRYKHLYVVSTTIVEYKPGAFVSRLRHNRDETPYSKNLLCLSNSSSSDGLTLQESHDIIKLIHLTQAPPSCTSHLLRDFPGTRAIEAPWMHFNKIKTNRLRLAAGIISLGGYAPASAALTRPIHIISGVTRWLPPLTYPTTSGTSNVALPHAAWVGCDTSDLHFVKHPTLRRITVKGLTRVTSLLRFVNNSPVELEDVTFVFTRPFEVQRLLLSARQCGPSRKSDKPYSQFNELPVEYNISRRSTRPDHSNPLRAQRERQVLEPDIDLCSHGMLLFRHPPGRLCLSIRSTNPRAPRKSGKAHDVDHRPHGLRCRVLRFEASSAPGSVWSAFCNNLPLGSLLACPVGAMLRRAVREAKLTVTGVTLTNVIANETWASGPPLGRRRLSGIALTGSLHCPRAVSPYCLLRSLICCEVLGYGAGCLLRLISSQSVLRVMLEARRTTLGIGSIIATKFTCLLTQVWQEAANHKCHGARIRMSLQSFGTGRTIASNNAPRRSSVELGHQPCHLRLAKFAVKDRPIAIYRPINANEWGYGIATKNTSFYHRRILAATGSFNTKSSGKGGVKSTDLELYRSSNKMSSVTRRSMLSTLALSKSLRRSRRIVQLAPRRPRHWNFSDVYSPLTIAGAYHSDKQTLALTSRSAYLVGRSVTLLTKRKDYDSLLEFAFLFDTALSALSSIPLSVWVRPKNLTRFSILNISYLLPCVNDVTAALSRTTNFGCILKSAIRYYRPDKHSPVNVAVYPDAERIRYRVNAERTLVTLRVGALCLYPTGFVDNAADVALYAYGVYVGGLEPRPTLYKIVSGQQAAIQNRAEIQTTARYVTVPSDVNDLGEATPALTKVNISVISVPLSRLSGSTGMKAPTFEAPTTGSESECSIRCPLVMKHGPWLENGFRSVASRAVAKSITFEEIGYKARTKANTFCRPSSCRLNFRSLLPWGFREPNSVNGSGVAARVLICSLKPPLNREEEKIRAVRGLLWSVYAKKARACRVGHSGSNLPMSDPDPRRLLVPDSRRDYCNYEDSKGIIPTTNYCARRVPPYRKTGPIARGSQPAVKDSPSMRPAATRSFKRHARAKQHAMVVLSKLLLDGGHCCTIPFRSYTKLSMAVRPGQVLSRAKMLFKILSFHEVLTTRLLHTFGLHSFTMEAYTVSLYSYAARVCSHDPLVGLYYGTKDTRIGPVRNLRLSLRPIRDILSRGHLVPTVPVTSLGRNKTGVNIRTFSSAPLYDGIEFGPHLSALACSSSHYVRPVGDIHRINAQYLIQRPQRQLPFNSETPVSQNELLTVNLYQYFTCEPIDHCPRTSDTSTLEDLMPLVPRPLPLSATMYTDALYTHVSKTLLSYSYDQHKPPTINALKGEGAPIWGPLSIDRVTRLASTVEILRLSDYESTFYPPTPDGDSYLTKLYKSMTPAQCVRLMLEPKRTEKTRGGNSGKLQSDAQSCADLQRTLSTKLRWRSEAAEDALTSPATDDIVFCPAHHLRIHPFSRYMQSLTSPKATRELKMGILPIPLPPYSRHCASYISISIIGHGKSLCPCGLATVAYPHRDDITNQVSSSVKNLVLQSMCPRILIQHLLSIVRVLVVNSTLCHWIVRYVYPKCDEYVSGDNIVACQSRLLLRVYSDELTASLAKMSSRFQTDLLGRVRGRLPAPELTVRAAGSPNEESTKLTSEHLVSPLLSCPQECFNRQNIHVLISFSQAYADSGVVQCQNGDSLHIFHSNSRRIVLGVMSFMLSAKVGHRNYGGSDDNADSHTHRGSRGDLSADFLVKSRGRVWRSVWYLLKNAISGLRCPSWIWSNTIAYQASRTIMSIGRLRQINQGQISPRESGGQTTSVQGVKNETGSICSKGFFIVWSSVEAPTFHDDMQARSSIGQAIIRCASELDARVRDVETVGSGADGLPSESSSTCSSFRKLMVGTPPPAVHETKVRKTSAMRARTDCLAQRGEPVKVLVSSPLGRNWTALISRGRYERHPTEPTKSSQLDETTRAMRQEIRKPSSVSCYQCSKEGHKHRACVSRLAWFINWVHNKYTMAPSLLHCGGRGVSRVALNTFKIECFLSPVLRARYASAERSLQHDLRVLSTPVLSNLHGDTEESCGIYSVALQGCLLYHSPKRQYSYPQSPTRVGRNLSKRVCKAQVRSARWIALQTNKYFCDDGSPSGLVFTARAAFFTEAFPHLALIQLKNTLLNVAHNSAVPRDDSDCHLSELSYPTARMSLPSRREPQDKVSWLADWKTPRIQTPYRQLVTSSYHKLPSVESAISLDCAHSLLYVGLKVAVDSSGWCVSYWHRHPDTPGTRTTDAARLGNTMSDTTNAKVWHPRKGLCSTNPLIVGPTTVLPSSVSRSSTRLHWEKMLLNWQSRAGDCRTTNVGPCRCAVNKGRIGIVLRTCHESAEAHSAQRITTRSSSRRPLGLTVNYHLSKSSLAWQYVAGEYIYFTEGLPLHRNRNHCYAKMSVEFITMLTECTSGENGKISTRAVFRTEYSYSSIGSNGKSEASGTRCGACRAKRAGLNASNVTDCLGPWAAALSNVVRFGVGLKVLLLSAWGVHAAKSQACPYLRQAGETNSAIWVEEVVVRPNESQIVGGCSNIWREVSLKSLRIYRYKSLPNTDRRLSDRMNADTYTSTPTLKYSGSTLARSIQGCVKKMQSTTQLRLRDPTLHSYLLQLELVLATPLPRM\n"
     ]
    }
   ],
   "source": [
    "# Create dict where k-v pair corresponds to RNA condon-amino acids for the purpose of protein creation.\n",
    "dict = {}\n",
    "with open('question_inputs/rna_codon_table.txt', 'r') as f:\n",
    "    for line in f:\n",
    "        k,v = line.split()\n",
    "        dict[k] = v\n",
    "        \n",
    "# input string\n",
    "s = 'AUGUGCCAUAUUGUUACCAGAUAUUCUGCACACCGCCCGGUACAUGCGGAACUUCAAUUUGCAAGCUCACCUACGCGCUCUUCCGACGGGGGGUACAAAAGCUUGAAGGCCAUGGGUACUACCGUUCUUCUCCGGUCGUUGCGGUCCAUAUCUUGCCGUACUUAUUAUAUUAUACCCUGCCGCUCGCCGAGGCUCGAGCACUUUCGCCUCAGCUGCAAGAUCUCGCUGUUCCACGCGCGUCAGACAAUGCUUAAGGUUACAGUAAUGACGUACCCCGCAAAAAUUGCGGCGUGCGCAAGGAGGGCUAAACUACCCUUCAUACUGUGCGGUGGUAUGGUGCCUAGCUGUUCCCGUGUCAUGAAAUACGGGUGUAUCCGCUACCGUGAGACGGGAAAACAACUCAAGUGCGUUAAGACGGCCCUACGUCCUACUAAUGAGAAUAGUUCUAUUGCCCGUGCGAAGUGCAAUGUCAAUGUGAAAGGUUGGCUACGGUAUCAGUUAUGCGCCGAUCAAGGUAUGGCGCAACUACGCGCCUGCCCUAUGGGUCGAAGGGGUCGAUCGGACAUGAGGUUGAUAGGACCUCGCGUGCUACUACCAUCGGGCGGCUGCAGCAAAAGUGGCUGGUACAAUCUUAGUCUAACUGACAGACCGCUGCCGGUCUCAUGUUUACCCAUGGGCUCGUUCAAGUCUGCAUCGUCGAAUAUUACAGGGGCGAGCCCCUGUAUGACUAUAAUGCCAAUUAUCAAAACACAGAAUUUUGCUCAUUGUAUGCCAAAAUUGACCCCGCGGAUAACUCAGCAGUUGAAACGACUUCUGCUGCCUCCGGUGUUUGCUCCCGCUAGGCCACGAGGCCAUUCACACGUGUAUCAGACCGGGAUUGCUUUCUGCACACCGUGGACACCUGAGCCCUACCAAUCGCGAUGUUGCCAUGGGUUGCUUGCAGUUCUAUUGAUAAGGCGCACAUUUACUUCCUUGCAACCGAGCCGCUUCCGUAAUAACAUAGCCGAGGUAUCACUCUCGGAGAGCAUUUGGGCAUAUUGCUACCACAUCGAGCCGUACUCACUUGAGCUAUACCCUGAAAUGGGUCAUGCAAGUACCUCCGUAAGUCGGCUCCGUCCCCGACAAGUAGCGGUGCGAGUAAGAGCAAACAGACGCAUAAGUCUCAGAGUCCAAACGCAAUGUGGGAAGUUGCCCACUGGUUUGUGGGUACCCAUCGGUAAAUCUAACGGAGACACUCUGCAACCGUGGGUUGAAGUAGAUAAAAAAUUCCUGGGUGAACGUACGGACACUGUAGAGACUUAUUUAGAAAUUUGCAUCGUCAGGCCGGAGGGCUUUUCUAGAUCUGUAGAGUUACAGAGAUGCGAUUUCAGGUUCUGGACGCAAGUUUCCACUCAAGUCACAAAGCCACCAGUGAAAGCUGCCGUAUGCAUCUACAAUGUCUCCCGGCGUUCAUACUUAUCGGCGCCUAGCAACCCGUCCCUCCCCCUUUUUAGCCAACGAUGCUGUAACUGGGGUUUCUCCACACACUCAGAAAUGCCGGGGGACGAUCUUUCUCCUUUGCUGUCGGUAAAACAUUUACACAUGGGGCUACGGGUACUGGAAGCAGGGUACCAAGACUCCAAACGUCACUAUUAUACUACAUGCCGGGAGUCUUCACUUUAUUUGCGAAUUGCGCUUGCGGUGGUAGCGCGGCGCAAGGUCCGGCAUCACAGCCAAGACGAUACUAAGAGUCUACCUUGGAACAACAGUCGAUAUAAACAUCUCUAUGUCGUAAGUACUACAAUUGUUGAGUACAAGCCGGGCGCCUUUGUAUCGAGACUCCGGCACAACCGCGACGAAACUCCAUACUCCAAGAACUUGCUCUGUCUAAGCAAUAGUUCAUCAUCGGAUGGAUUGACCCUACAGGAGAGCCACGACAUUAUAAAACUGAUCCACUUGACGCAGGCUCCUCCCUCUUGCACUAGUCACCUACUAAGAGACUUUCCAGGGACACGGGCGAUAGAAGCACCAUGGAUGCAUUUCAAUAAGAUCAAGACCAACAGGCUGAGGCUCGCGGCAGGCAUCAUUUCCCUUGGCGGCUACGCACCUGCAAGCGCGGCGCUUACACGUCCGAUUCAUAUUAUUUCGGGCGUGACGCGUUGGCUGCCCCCCCUAACAUAUCCAACCACAAGUGGAACAUCCAAUGUGGCACUCCCACACGCUGCGUGGGUGGGAUGUGAUACUUCAGACCUACAUUUUGUUAAACAUCCAACUCUACGUCGAAUUACGGUUAAAGGGCUCACAAGAGUGACUUCACUAUUAAGGUUUGUAAAUAAUUCUCCCGUGGAACUGGAAGACGUUACCUUUGUGUUUACGCGUCCAUUCGAAGUCCAGCGACUCUUGCUAAGCGCACGUCAAUGCGGCCCGUCACGUAAAUCUGACAAGCCGUACUCACAGUUCAAUGAGCUACCAGUGGAAUACAACAUAUCCCGGAGGAGCACACGCCCUGAUCACUCCAAUCCUUUACGGGCACAGCGGGAACGUCAAGUACUUGAACCGGACAUUGAUCUUUGUAGUCAUGGCAUGUUGUUAUUUCGUCAUCCGCCAGGUCGUUUAUGCCUGAGCAUUCGAUCGACCAAUCCGCGCGCUCCCAGGAAGUCAGGAAAAGCCCACGAUGUCGAUCAUCGUCCUCACGGACUCAGAUGCCGCGUGCUGAGAUUCGAAGCGAGUUCCGCUCCAGGUAGUGUGUGGUCCGCCUUCUGCAAUAAUCUGCCCUUAGGGUCUUUGCUCGCAUGCCCGGUGGGUGCUAUGUUAAGAAGGGCGGUUCGAGAAGCGAAGCUAACGGUGACAGGUGUUACUUUGACAAACGUGAUAGCAAACGAGACGUGGGCCUCUGGACCCCCGCUUGGUAGGCGACGUCUCUCCGGCAUUGCGCUUACGGGUUCCCUUCACUGCCCUCGUGCAGUCUCACCCUAUUGCCUUCUGAGGUCUCUCAUAUGUUGUGAAGUACUAGGUUAUGGGGCCGGCUGCUUACUUCGUUUGAUCAGCAGCCAGUCGGUGUUGAGAGUUAUGCUAGAGGCGCGACGCACCACACUCGGGAUCGGCAGUAUUAUAGCAACAAAAUUCACGUGCUUGUUAACCCAAGUGUGGCAAGAAGCAGCCAACCACAAGUGCCACGGCGCUAGGAUUAGGAUGUCUUUACAGUCUUUUGGAACCGGUCGUACGAUAGCGAGUAAUAACGCCCCCCGAAGAUCGUCGGUUGAGUUGGGUCACCAGCCAUGUCAUCUCCGGCUCGCUAAAUUUGCUGUAAAAGAUCGUCCGAUAGCCAUAUAUAGGCCAAUCAACGCCAAUGAAUGGGGCUACGGUAUUGCAACAAAGAAUACUAGUUUCUACCAUCGUAGGAUACUGGCGGCCACCGGGUCCUUCAAUACUAAAAGUUCGGGGAAAGGGGGGGUUAAGUCUACUGACCUAGAACUAUAUCGGAGCUCGAAUAAGAUGUCGUCGGUCACGCGGCGCAGCAUGCUAUCUACGUUAGCCCUAUCGAAAUCGCUGCGAAGGUCGCGUCGAAUCGUUCAGCUCGCGCCUCGUCGACCCCGGCAUUGGAACUUUUCCGAUGUAUAUUCGCCGCUGACAAUUGCGGGCGCGUAUCAUAGCGACAAGCAAACUCUAGCACUGACUUCUCGGUCGGCAUACUUGGUAGGUAGAUCAGUAACGCUCCUUACUAAACGUAAAGAUUAUGAUUCACUAUUAGAGUUUGCGUUUCUAUUUGAUACUGCGCUUUCCGCUCUAUCAUCUAUCCCCCUCUCCGUUUGGGUUCGUCCAAAAAACCUCACAAGAUUCUCUAUAUUGAAUAUCUCAUAUCUGCUUCCGUGUGUGAAUGACGUCACAGCAGCACUGUCCCGCACAACAAAUUUCGGUUGUAUCCUCAAAUCGGCGAUCCGCUAUUACCGGCCGGAUAAACACAGCCCUGUUAAUGUAGCCGUUUAUCCAGACGCAGAGCGGAUCAGAUAUCGAGUAAACGCGGAACGUACCCUAGUAACCCUUCGUGUAGGUGCUUUAUGCCUUUACCCUACUGGCUUUGUCGAUAAUGCAGCUGAUGUUGCUCUCUAUGCUUACGGGGUCUACGUCGGAGGCUUAGAACCUCGUCCUACACUUUAUAAAAUAGUAUCCGGGCAACAAGCUGCUAUCCAGAACAGGGCCGAGAUACAGACAACCGCUAGGUACGUAACAGUACCAAGCGACGUUAACGAUCUUGGAGAAGCUACACCGGCCUUAACUAAGGUAAAUAUCUCAGUGAUAUCUGUUCCACUGUCACGUUUGUCGGGCUCCACCGGAAUGAAAGCCCCUACAUUUGAAGCACCUACAACUGGAUCCGAAAGCGAGUGCUCUAUACGUUGCCCUUUGGUUAUGAAACAUGGGCCAUGGCUUGAGAAUGGUUUCCGUAGUGUCGCCUCUAGGGCCGUCGCAAAAUCCAUUACGUUUGAAGAGAUCGGGUACAAGGCGAGGACCAAAGCCAAUACGUUCUGUCGGCCAAGCUCCUGUAGACUCAAUUUUCGUUCGCUGCUACCUUGGGGGUUUAGAGAACCGAAUUCGGUCAAUGGGUCGGGUGUAGCAGCGCGAGUGCUAAUAUGUAGUCUGAAGCCGCCGCUCAAUCGGGAAGAAGAAAAGAUUCGGGCAGUAAGGGGACUGUUAUGGUCGGUAUAUGCCAAAAAGGCCCGAGCAUGCCGUGUCGGGCAUUCAGGCAGCAAUCUGCCAAUGUCGGACCCUGACCCUAGAAGGUUGCUAGUCCCGGAUAGUCGCCGGGAUUAUUGUAAUUACGAAGAUUCAAAAGGCAUAAUCCCGACCACGAAUUAUUGCGCCCGCAGAGUACCACCCUACCGCAAAACAGGGCCAAUAGCGCGUGGAAGCCAACCGGCAGUGAAAGAUUCCCCUUCAAUGCGACCAGCCGCAACUCGCUCAUUCAAAAGACACGCCCGUGCCAAACAGCACGCUAUGGUAGUACUCUCUAAACUAUUACUUGAUGGUGGGCACUGCUGCACGAUUCCCUUCAGGUCGUAUACAAAAUUGUCGAUGGCUGUUAGGCCGGGACAAGUCCUUUCCAGGGCAAAGAUGUUGUUUAAGAUCCUAUCUUUUCAUGAGGUACUGACUACCCGGCUGCUGCACACUUUUGGUUUGCAUUCGUUUACAAUGGAAGCCUACACGGUGUCCCUUUAUUCUUAUGCGGCGCGUGUGUGCUCCCAUGACCCUCUAGUGGGCUUGUACUACGGCACUAAAGAUACCCGUAUCGGGCCUGUCCGUAACUUGCGGUUAAGUCUGAGGCCAAUACGAGAUAUCCUUUCGCGUGGACACCUGGUCCCCACAGUUCCCGUCACAAGCCUGGGGAGGAAUAAAACCGGGGUAAAUAUUCGCACGUUUUCAUCGGCGCCCUUGUAUGACGGAAUUGAGUUUGGCCCACAUCUCAGCGCCCUCGCUUGUUCGUCAAGCCACUACGUUAGACCCGUGGGGGAUAUCCAUCGUAUAAAUGCUCAAUAUUUGAUUCAGCGUCCUCAGCGCCAGUUGCCGUUCAAUUCAGAAACCCCGGUAUCGCAGAACGAACUCCUCACCGUCAAUCUGUAUCAAUAUUUCACUUGCGAGCCCAUAGACCAUUGUCCGCGCACUUCCGAUACCUCGACGCUUGAGGACCUAAUGCCACUAGUGCCUCGACCUCUUCCGCUAUCGGCUACCAUGUACACGGACGCGCUCUAUACCCAUGUGUCCAAAACUCUACUAAGCUAUAGUUACGACCAACACAAACCUCCGACCAUAAAUGCGCUAAAGGGCGAGGGCGCGCCAAUAUGGGGGCCCCUGUCAAUCGAUAGGGUUACUCGACUGGCGUCAACUGUGGAGAUUCUUAGAUUAAGUGACUAUGAAUCAACCUUUUACCCGCCUACGCCGGACGGCGAUUCAUACUUGACGAAGUUAUACAAGUCGAUGACACCCGCCCAGUGCGUGAGACUCAUGCUUGAGCCAAAGCGUACCGAAAAAACUCGGGGGGGGAAUUCCGGUAAGUUGCAGAGCGAUGCACAAUCAUGUGCGGACCUUCAAAGGACUCUUUCGACCAAAUUAAGAUGGCGCAGCGAGGCGGCAGAGGAUGCACUUACUAGUCCUGCGACUGACGAUAUCGUUUUCUGCCCAGCCCAUCACCUUCGAAUACAUCCCUUCUCCCGCUAUAUGCAGUCCCUUACCUCUCCGAAGGCUACCAGAGAACUCAAGAUGGGGAUACUACCAAUACCUCUCCCGCCUUACUCCCGCCACUGUGCGUCUUACAUAUCGAUUUCUAUAAUAGGGCAUGGGAAAUCGCUAUGCCCAUGCGGACUUGCCACCGUGGCCUACCCUCACCGUGACGAUAUCACUAAUCAAGUCUCAAGCUCAGUUAAAAACUUGGUUUUACAAUCAAUGUGCCCUCGUAUUCUAAUUCAGCACCUUCUGUCAAUUGUUCGUGUCCUUGUGGUUAAUAGCACUUUAUGUCAUUGGAUAGUACGAUAUGUCUACCCCAAAUGCGACGAAUAUGUGUCAGGUGACAACAUCGUGGCUUGCCAAUCCCGACUCCUCCUGCGGGUCUACAGCGACGAGCUCACAGCUUCCCUAGCAAAAAUGAGUUCCCGCUUUCAAACUGAUUUGCUGGGCCGGGUGAGAGGGAGACUGCCGGCACCAGAAUUAACUGUACGGGCUGCGGGGAGUCCUAACGAAGAAUCAACCAAACUUACCAGCGAGCACCUUGUUUCCCCCUUGCUAUCAUGCCCGCAAGAAUGUUUUAACAGGCAGAACAUACAUGUACUCAUAUCCUUUAGCCAGGCGUAUGCUGACUCAGGAGUCGUCCAAUGCCAGAAUGGUGAUAGUUUACAUAUCUUCCACUCAAACAGCCGCAGAAUCGUCCUGGGCGUUAUGUCAUUCAUGCUAUCGGCAAAGGUUGGGCACCGGAAUUAUGGUGGAAGUGAUGACAAUGCAGACAGCCACACACAUCGCGGUAGCCGUGGGGAUUUAUCGGCUGACUUCUUAGUUAAGUCGCGCGGUCGUGUAUGGAGAAGUGUAUGGUAUCUCCUGAAGAAUGCAAUCUCCGGUUUAAGAUGUCCCUCCUGGAUAUGGAGUAAUACCAUCGCUUAUCAGGCGAGUAGGACGAUAAUGUCAAUUGGGCGUCUCCGACAGAUUAACCAGGGUCAGAUCUCUCCACGUGAGAGUGGGGGUCAAACGACCAGUGUGCAGGGCGUUAAAAAUGAAACAGGCUCCAUAUGCUCCAAAGGAUUCUUUAUCGUAUGGUCGAGUGUUGAGGCUCCAACGUUCCACGAUGAUAUGCAAGCUAGGUCUUCUAUUGGGCAAGCAAUUAUACGCUGUGCCAGUGAGCUGGAUGCGCGCGUACGAGAUGUUGAAACGGUGGGAUCCGGGGCAGACGGGCUCCCAUCGGAGAGCUCUUCAACAUGUUCAAGCUUUCGGAAGCUGAUGGUUGGUACUCCGCCACCCGCGGUGCACGAGACCAAAGUAAGGAAAACCUCUGCGAUGCGUGCCAGGACAGAUUGCUUAGCACAGCGCGGGGAGCCAGUGAAGGUCCUGGUAUCGUCUCCAUUAGGUAGAAACUGGACUGCUCUAAUAUCUAGAGGGCGCUAUGAGAGGCAUCCAACUGAGCCUACGAAAUCUAGCCAGCUUGAUGAGACGACGAGAGCUAUGCGACAAGAGAUACGCAAGCCCAGUUCCGUUAGUUGCUAUCAAUGUUCGAAAGAGGGACACAAGCAUAGAGCGUGCGUUUCACGCCUGGCUUGGUUUAUAAAUUGGGUUCACAACAAGUACACGAUGGCACCUAGCCUCUUGCAUUGCGGAGGACGUGGCGUUAGCCGCGUCGCACUAAACACUUUUAAAAUCGAAUGUUUCUUAAGCCCAGUACUGCGAGCACGUUAUGCCAGUGCAGAACGUUCCCUACAGCAUGAUUUGAGAGUCCUCAGUACUCCUGUCCUGUCAAAUUUGCACGGGGACACCGAAGAAAGUUGCGGGAUUUACAGCGUGGCGUUACAGGGUUGUCUUCUCUACCACUCACCCAAGAGACAGUAUUCCUAUCCGCAAUCACCUACGCGUGUAGGCAGAAACCUCAGCAAACGCGUGUGCAAGGCGCAAGUUCGUUCCGCACGAUGGAUUGCCUUGCAAACGAAUAAAUACUUCUGCGAUGACGGGAGCCCCUCAGGCCUUGUCUUCACGGCACGAGCAGCCUUCUUUACCGAGGCGUUUCCCCAUCUAGCCCUCAUCCAGCUUAAGAAUACAUUGCUUAAUGUUGCACAUAACAGCGCGGUACCCCGAGAUGAUUCCGACUGUCAUCUCUCGGAAUUGUCGUACCCAACGGCCCGGAUGUCGUUGCCCUCAAGAAGGGAGCCGCAAGACAAGGUCAGUUGGUUAGCAGAUUGGAAGACACCUCGCAUUCAAACACCUUAUCGUCAACUCGUUACGUCUAGUUACCAUAAAUUACCCUCCGUUGAGUCGGCAAUAAGUCUCGACUGUGCCCAUUCUCUGCUUUAUGUGGGUCUUAAGGUUGCGGUCGAUUCGUCGGGAUGGUGCGUAUCUUAUUGGCACAGACACCCUGACACGCCAGGAACACGAACUACAGAUGCUGCUAGGCUAGGGAACACUAUGAGUGACACAACCAACGCGAAAGUUUGGCAUCCGCGAAAAGGGCUUUGCUCUACGAACCCACUUAUUGUAGGCCCGACAACGGUGCUCCCUUCGUCCGUCUCUAGAUCCUCGACUCGUUUGCAUUGGGAGAAGAUGUUACUGAACUGGCAAUCCCGUGCAGGGGAUUGUCGGACGACAAACGUUGGUCCAUGUAGAUGCGCCGUUAACAAGGGACGUAUCGGGAUAGUUCUGCGGACAUGCCACGAAUCCGCCGAAGCACAUUCUGCUCAACGAAUUACUACAAGAUCCUCCUCUCGGCGUCCGCUAGGUCUCACCGUUAACUACCACCUUAGCAAAAGCAGCCUGGCCUGGCAGUAUGUGGCAGGGGAAUAUAUCUACUUUACCGAGGGUUUACCCUUGCAUAGAAAUCGGAAUCAUUGCUAUGCUAAAAUGAGCGUGGAGUUUAUCACGAUGUUGACUGAGUGCACCAGUGGCGAGAACGGCAAGAUCUCUACUCGUGCAGUCUUCCGGACAGAGUAUUCUUACUCAUCUAUUGGCUCGAACGGUAAGAGUGAAGCUAGCGGUACCAGGUGUGGGGCUUGUCGAGCCAAGAGGGCCGGUCUCAAUGCCUCAAACGUGACUGAUUGCCUAGGUCCUUGGGCAGCGGCUUUAUCGAAUGUGGUGCGAUUUGGCGUGGGACUGAAAGUUUUAUUGCUGUCUGCAUGGGGCGUCCACGCGGCUAAGUCACAGGCGUGUCCUUAUCUGAGACAGGCUGGAGAGACUAACUCGGCCAUAUGGGUUGAGGAGGUCGUUGUGAGGCCGAAUGAAAGUCAAAUCGUCGGUGGUUGCUCGAACAUUUGGAGAGAGGUAAGUUUGAAAAGUCUUCGCAUCUACAGGUACAAGUCUUUGCCAAACACCGAUCGGAGACUAUCCGAUCGCAUGAACGCCGAUACUUACACUUCCACGCCAACCCUUAAAUAUUCGGGGAGUACCCUUGCACGAUCAAUCCAAGGUUGUGUGAAGAAAAUGCAGUCGACUACCCAACUCCGUCUCCGCGAUCCGACUUUGCACAGUUACCUGCUGCAGCUUGAACUAGUUCUUGCGACUCCUCUCCCGCGCAUGUAA'\n",
    "output = ''\n",
    "i = 0\n",
    "while i < len(s):\n",
    "    codon = s[i:i+3]\n",
    "    amino_acid = dict[codon]\n",
    "    if amino_acid == 'Stop':\n",
    "        break\n",
    "    else:\n",
    "        output += amino_acid\n",
    "    i += 3\n",
    "print(output)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Finding a Motif in DNA\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If the same interval of DNA in the genomes of two different organisms is found together, it is suggestive that the interval has the same function in both organisms. \n",
    "\n",
    "A **motif** is such a commonly shared interval of DNA. A common task is to search for an organism's genome for a known motif. \n",
    "\n",
    "Genomes are riddled with intervals of DNA that occur multiple times (perhaps with only slight modifications) called **repeats**. The occur too often to be simply random chance. Thus the language of DNA is quite powerful.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem\n",
    "\n",
    "Given two string s and t, report the starting index of all occurances of t in s, using 1-based indexing. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[3, 10, 17, 56, 68, 126, 133, 215, 222, 233, 264, 272, 331, 379, 394, 401, 408, 432, 470, 486, 493, 501, 508, 637, 681, 688, 695, 712, 762, 795, 803, 810, 825, 847, 874, 881, 891, 898, 905, 920, 927, 952]\n"
     ]
    }
   ],
   "source": [
    "s = \"ATGGCACGAGGCACGAGGCACGAGGGCGGCACGATTCAGGCACGAAGGCACGACTGGCACGAGGCCGGGCACGAGGCACGAAGGCACGATGGCACGAAGGCACGACGGCACGAAAGAGGCACGAAGGCACGAGGCACGAGGCACGACCGAGGCACGAATCGGCACGATCCCTTTTAGGCACGATGAGCTCGGCACGATCCGCCAAGGCACGACTGGCACGAGGCACGAGGACGGCACGAGGCACGATTCCGGCACGATACCTGGGCACGAGGGCACGAGGTACACGACGGCACGAAACGGCACGACTACCGGCACGAAGGCACGACTACGGGCACGAGGCACGATGGTGGCACGACTGCGGCACGACGGCGGCACGACGGCACGAGGCACGAAGGCACGAGGCACGAGGCACGAGGCACGAGCCTTTCTAAGGCACGAGGCACGAAGCGTCAATTAGGCACGACAGCGCGGCACGAGGGCACGATGGCACGAGGCACGAGGGCACGAGGCACGAGGCCCACGGCACGAAATTGGGCACGATCGGCACGAGAGCAGGGCACGAGTCGAGGCACGATCGGGCACGAGCAGGCACGAGACGGCACGATTCTGGCACGAGTGGCACGAAGGCACGATCTGGGCACGAGGCACGATAAGGCACGATGGCACGAAGGGGCACGACGGGCACGAGGCACGAGGCACGAGGCACGAGATGGCACGAGGGCACGAAAGGCACGATCCCGTCACGGCACGACAGGCACGATGGCACGAGGCACGATCTGAGTGGCACGATCGGCGGCACGAGGGCACGAGGCACGAGGCACGACGGCACGAGGCACGACTGATGTGGGCACGAGGCACGATAGTCGGCACGACGGCACGAGGCACGAGGGGGCACGAGGCACGAGGCACGAGGCACGACGGCACGAGGCACGAGGCACGAGCGGCACGAACGGCACGAGGCACGA\"\n",
    "t = \"GGCACGAGG\"\n",
    "\n",
    "index_loc = []\n",
    "\n",
    "for i in range(len(s) - len(t)):\n",
    "    if s[i:i+len(t)] == t:\n",
    "        index_loc.append(i+1)\n",
    "        \n",
    "print(index_loc)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Consensus and Profile\n",
    "\n",
    "We can count the point mutations between two strings, but for multiple strings (of equal length) we can do more. For several homologous strands, we can find an average-case strand to represent the most likely common ancestor of the given strands. \n",
    "\n",
    "For each index of the the strings, by taking the largest number A,C,G, and T bases as a representative of all the strings forms the **concensus string**. The **profile** of the strings (all of length n) is the 4xn matrix where entry $1,j$ denotes the count of A in all strings at position j, $2,j$ the count of C in all string at position j, and so on.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Problem \n",
    "\n",
    "Given a collection of DNA strings of equal length, all in FASTA format, output the concnsus string along with the resulting profile."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Helper function: read FASTAf file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_FASTA_file(file):\n",
    "    # preprocessing\n",
    "    dna_list = file.readlines()\n",
    "\n",
    "    dict = {}\n",
    "\n",
    "    curr_key = ''\n",
    "    curr_string = ''\n",
    "\n",
    "    for item in dna_list:\n",
    "        # Line contains label\n",
    "        if item[0] == '>':\n",
    "            # store current_key with current_string\n",
    "            if len(curr_string) != 0:\n",
    "                dict[curr_key.replace('\\n','')] = curr_string.replace('\\n','')\n",
    "                curr_key = ''\n",
    "                curr_string = ''\n",
    "\n",
    "            curr_key = item\n",
    "        # line contains piece of string. \n",
    "        else:\n",
    "            curr_string = curr_string + item\n",
    "\n",
    "    # Add last pair to dict.\n",
    "    dict[curr_key.replace('\\n', '')] = curr_string.replace('\\n', '')\n",
    "\n",
    "    return dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'>Rosalind_1': 'ATCCAGCT', '>Rosalind_2': 'GGGCAACT', '>Rosalind_3': 'ATGGATCT', '>Rosalind_4': 'AAGCAACC', '>Rosalind_5': 'TTGGAACT', '>Rosalind_6': 'ATGCCATT', '>Rosalind_7': 'ATGGCACT'}\n"
     ]
    }
   ],
   "source": [
    "print(read_FASTA_file(open('question_inputs/consensus.txt', 'r')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "can only concatenate str (not \"list\") to str",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-114-9312b2d4c824>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     43\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     44\u001b[0m \u001b[0moutput_file\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mave_string\u001b[0m \u001b[1;33m+\u001b[0m\u001b[1;34m'\\n'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 45\u001b[1;33m \u001b[0moutput_file\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'A: '\u001b[0m \u001b[1;33m+\u001b[0m \u001b[1;33m[\u001b[0m\u001b[0mP\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m,\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mn\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m]\u001b[0m \u001b[1;33m+\u001b[0m\u001b[1;34m'\\n'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     46\u001b[0m \u001b[0moutput_file\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'C: '\u001b[0m \u001b[1;33m+\u001b[0m \u001b[0mstr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mP\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m+\u001b[0m\u001b[1;34m'\\n'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     47\u001b[0m \u001b[0moutput_file\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'G: '\u001b[0m \u001b[1;33m+\u001b[0m \u001b[0mstr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mP\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m2\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m+\u001b[0m\u001b[1;34m'\\n'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mTypeError\u001b[0m: can only concatenate str (not \"list\") to str"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "\n",
    "# Get contents of FASTA file\n",
    "dict = read_FASTA_file(open('question_inputs/consensus.txt', 'r'))\n",
    "\n",
    "output_file = open('question_outputs/consensus.txt', 'w')\n",
    "\n",
    "# create profile matrix\n",
    "\"\"\"\n",
    "row 0 -> A\n",
    "row 1 -> C\n",
    "row 2 -> G\n",
    "row 3 -> T\n",
    "\"\"\"\n",
    "list_strings = list(dict.values())\n",
    "P = np.zeros(shape=(4, len(list_strings[0])))\n",
    "P = P.astype(int)\n",
    "\n",
    "# Add values to profile matrix.\n",
    "for string in list_strings:\n",
    "    for i,char in enumerate(string):\n",
    "        if char == 'A':\n",
    "            P[0][i] += 1\n",
    "        elif char == 'C':\n",
    "            P[1][i] += 1\n",
    "        elif char == 'G':\n",
    "            P[2][i] += 1\n",
    "        else:\n",
    "            P[3][i] += 1\n",
    "\n",
    "# Output results, print to file. \n",
    "ave_string = \"\"\n",
    "for i in  range(len(P[0])):\n",
    "    idx = np.argmax(P[:,i])\n",
    "    if idx == 0:\n",
    "        ave_string += \"A\"\n",
    "    elif idx == 1:\n",
    "        ave_string += \"C\"\n",
    "    elif idx == 2:\n",
    "        ave_string += \"G\"\n",
    "    else:\n",
    "        ave_string += \"T\"\n",
    "\n",
    "output_file.write(ave_string +'\\n')\n",
    "output_file.write('A: ' + str(P[0]) +'\\n')\n",
    "output_file.write('C: ' + str(P[1]) +'\\n')\n",
    "output_file.write('G: ' + str(P[2]) +'\\n')\n",
    "output_file.write('T: ' + str(P[3]) +'\\n')\n",
    "\n",
    "output_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
